{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"lab13_protDCASBM.ipynb","provenance":[],"collapsed_sections":[],"toc_visible":true},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"WutUwTFo2e4j"},"source":["#**Lab.13 | IBM3202 (2021)** ‚Äì Combining Coevolutionary Analysis of Sequence Information and Structure-based Models to Predict Protein Structures"]},{"cell_type":"markdown","metadata":{"id":"FQMdCpigVK-X"},"source":["###Theoretical aspects\n","\n","The conservation of the three-dimensional structure among homologous proteins imposes constrains on sequence variability. Even when the sequence identity between homologous members from a given protein family can be even below the so-called twilight zone (20-30% sequence identity), their structures often exhibit **interactions between residue pairs that, when mapped onto their protein sequence, occupy equivalent positions among all proteins**. While the chemical nature of the interaction between these residue pairs can vary from one protein to another, the physical contact between these residues is **constrained in the final structure**.\n","\n","Based on this, it was suggested that the **statistical analysis of correlated amino acid substitution patterns at different sequence positions of a protein family** could be exploited to **infer spatial contacts within the tertiary protein structure**. The hypothesis behind it is fairly simple: if two residues are interacting in a protein structure and a destabilizing mutations occurs in one of the residues, **a compensatory mutation at an specific residue position in a sequence preserves the stability of the protein architecture and function (FIG 1).**  \n","\n"]},{"cell_type":"markdown","metadata":{"id":"P8b4ES6EHX9Z"},"source":["Thanks to the explosion of genomic sequencing, which has enabled the data deposition of **thousands of protein sequences** in publicly available databases, these statistical signatures left by protein evolution can be ascertained to predict which residue pairs should be in spatial proximity in the native functional protein fold."]},{"cell_type":"markdown","metadata":{"id":"xaWvTKKCDqfs"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/sbmdca_01.png'/ width=700>\n","<figcaption><b>FIGURE 1. </b>Schematic representation of the residue pair contacts that can be inferred from the coevolutionary analysis of linear protein sequences.<br>The left side of the image represents an alignment of multiple sequences of the same protein from many organisms. The boxed columns indicate two positions in this alignment whose residues are always complementary to each other, in spite of the mutations occurring in each position. This pattern suggest that these two positions likely form a physical interaction crucial for the stability of the protein structure, as shown on the right side.<br>\n","Cartoon by Sergey Ovchinikov, <i><a href=\"http://site.solab.org/home\">http://site.solab.org/home</a></i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"BfShtJPUa1gY"},"source":["These statistically inferred physical interactions are **a natural partner for Structure-based Models (SBM)**. The canonical SBMs **explicitly include a native bias in the potential energy function**, by treating all residue pairs **in contact in the native structure (FIG 2)** (given a distance cut-off) as **attractive interactions (FIG 3)**. All non-native interactions (i.e. those residue pairs with distances that fall outside the distance cut-off) are treated as repulsive. This condition is consistent with the notion of a sequence of a protein without highly frustrated interactions, or so-called **funneled folding landscape**."]},{"cell_type":"markdown","metadata":{"id":"J2eGQeq9dOtn"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/smog_01.png'/>\n","<figcaption><b>FIGURE 2.</b> Using a distance-based criteria to extract native contacts from a protein structure. The left panel shows the proximity of the nearest atomic contact for each residue pair in the ribosomal protein S6 up to a maximum of 1.5 nm, whereas the right panel shows the resulting coarse-grained native contact map upon filtering by a distance cutoff of 0.6 nm and a minimal sequence separation of 3 residues between interacting residues <br> Noel JK & Onuchic JN (2012) <i> Computational Modeling of Biological Systems, 31-54</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"XYsrb6vrhQu3"},"source":["These attractive interactions are often model either as Lennard-Jones or Gaussian potentials, the latter corresponding to:\n","\n","<center>\n","$C_G(r_{ij},r^{ij}_0) = \\left(1+(\\frac{\\sigma_{NC}}{r_{ij}})^{12}\\right)\\left(1+G(r_{ij},r^{ij}_0)\\right)-1;$\n","\n","where $G(r_{ij},r^{ij}_0) = -\\exp \\left[-(r_{ij}-r^{ij}_0)^2(2\\sigma^2)\\right]$\n","\n","<figure>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/sbmdca_03.png' width=500/>\n","<figcaption><b>FIGURE 3.</b> Comparison of Lennard‚ÄìJones (LJ) and Gaussian contact potentials. Black curves show LJ contact potentials with minima at 6√Ö and 10√Ö. The Gaussian contact potential shown in red has an excluded volume Ùè∞É$NC$ that can be set independently of the location of the minimum. The dotted red line shows how the Gaussian contact would change as another minimum at 10√Ö is added <br> Noel JK & Onuchic JN (2012) <i> Computational Modeling of Biological Systems, 31-54</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"tujZVSladJGZ"},"source":["SBM further reduce the complexity of the biomolecular system by using a **coarse-grained approximation**, in which each residue is replaced by a bead centered at the alpha carbon coordinates. The decrease in granularity of the  system, along with its simplified energy function, allows for efficiently simulating protein folding with minimal computational resources.\n","\n","In this context, it becomes straightforward to **replace the distance-based native contacts extracted from an experimental structure by those predicted by coevolutionary analysis of protein sequences**, in order to use SBM along with DCA for **protein structure prediction**."]},{"cell_type":"markdown","metadata":{"id":"eiopP7EdVjqt"},"source":["##Experimental Overview"]},{"cell_type":"markdown","metadata":{"id":"xkpCp93eQsUS"},"source":["In this tutorial, we will exemplify how we can **infer native contacts from the coevolutionary analysis of protein sequence information using Direct Coupling Analysis (DCA)**, and how we can then **incorporate this information in Structure-Based Models (SBM)** to predict protein structures.\n","\n","We will particularly focus on comparing the inferred and experimental residue pair interactions of a **myohemerythrin** from the marine worm *Themiste hennahi*. Hemerythrin proteins ([**Pfam family PF01814**](https://pfam.xfam.org/family/PF01814)) are responsible for oxygen binding and transport in marine invertebrates and, contrary to expectations, they do not contain an heme group (as in hemoglobin). These proteins are quite small (less than 150 amino acid residues), which is a perfect protein size for our short tutorial.\n","\n","For the coevolutionary analysis of linear protein sequences, we will employ the **DCA** method. The DCA scores deliver quantitative information about the existence of physical contacts in a three-dimensional structure of a biomolecule, with top-scoring pairs accurately predicting native contacts that can be observed in experimentally solved protein structures.\n","\n","The top scoring residue-pair interactions inferred by DCA will be then incorporated as **Gaussian Interactions** (FIG 3) into **SBM** models, and combined with secondary structure parameters generated based on sequence predictions using **Jpred4**, for folding and predicting the three-dimensional structure of myohemerythrin.\n","\n","The following tutorial is based on three different sources of information that we highly recommend to further explore:\n","\n","- A tutorial on pyDCA by Mehari B. Zerihun (available in [this GitHub](https://github.com/KIT-MBS/pydca)), which is associated with the publication:\n","  -  Zerihun, M. B., Pucci, F., Peter, E. K., & Schug, A. (2020). pydca v1. 0: a comprehensive software for Direct Coupling Analysis of RNA and Protein Sequences. *Bioinformatics, 36*(7), 2264-2265.\n","- A tutorial on SBM by Jeff K. Noel and Paul C. Whitford (available in the [Structure-based Models On GROMACS [SMOG] server](https://smog-server.org)), with its usage and a general overview extensively described in the publication:\n","  - Noel, J. K., & Onuchic, J. N. (2012). The many faces of structure-based potentials: from protein folding landscapes to structural characterization of complex biomolecules. In *Computational Modeling of Biological Systems*, 31-54; Springer, Boston, MA.\n","- A tutorial and Python scripts on the combination of SBM and DCA by Ricardo N. Dos Santos (available in the [Morcos Lab website](http://morcoslaboratory.org)), which is associated with the publication:\n","  -  Dos Santos, R. N., Jiang, X., Mart√≠nez, L., & Morcos, F. (2019). Coevolutionary Signals and Structure-Based Models for the Prediction of Protein Native Conformations. *Methods Mol Biol, 1851*, 83-103"]},{"cell_type":"markdown","metadata":{"id":"buTRd47h4QfK"},"source":["#Part 0. Downloading and Installing the required software"]},{"cell_type":"markdown","metadata":{"id":"KSAoowL_4XIV"},"source":["We first must install several pieces of software to perform this tutorial. Namely:\n","\n","- **biopython** for manipulation of the PDB files\n","- **py3Dmol** for visualization of the protein structure.\n","- **HMMER** for aligning multiple protein sequences using profile Hidden Markov Model (HMM) of position-specific conservation.\n","- **pyDCA** for predicting native contacts through Direct Coupling Analysis.\n","- **SBM-enhanced version of GROMACS** containing Gaussian contact potentials.\n","\n","**WARNING‚ö†Ô∏è:** pyDCA requires more recent version of several python modules, and thus a **runtime restart** will be requested. If you are requested to perform a runtime restart, please do!\n"]},{"cell_type":"markdown","metadata":{"id":"7t7-UJnjOm8D"},"source":["1. We will first install **HMMER** from the Advanced Package Tool (APT)"]},{"cell_type":"code","metadata":{"id":"ba9bbMFk5ol3"},"source":["!apt-get install hmmer"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"x8zFzFh0OpPX"},"source":["2. We will then install **py3Dmol**, **pydca** and **biopython** simultaneously as they are all available through `pip`"]},{"cell_type":"code","metadata":{"id":"My3ObMLDY9G9"},"source":["!pip install pydca biopython py3Dmol"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"acZ-vteQyF7Y"},"source":["3. Lastly, we will install an **SBM-enhanced version of GROMACS** from the developers of [SMOG server](https://smog-server.org/SBMextension.html), which has been pre-compiled for the purposes of this tutorial.\n","\n"]},{"cell_type":"code","metadata":{"id":"YsuDg8S4xF7P"},"source":["# Download and unzip the compressed folder of SBM-enhanced GROMACS\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/software/gromacs_sbm.tar.gz\n","!tar xzf gromacs_sbm.tar.gz"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"BRKHqTOyyfC8"},"source":["Once these software installation processes are completed, we are ready to perform our experiments"]},{"cell_type":"markdown","metadata":{"id":"9GjXpNlbW-a3"},"source":["#Part I ‚Äì Examine the three-dimensional structure of *T. hennahi* myohemerythrin"]},{"cell_type":"markdown","metadata":{"id":"MoxY4Y_YXYFd"},"source":["We will first download and visualize the solved structure of *T. hennahi* myohemerythrin (PDB 2mhr) through biopython and py3Dmol.\n","\n","When working with protein structures solved by X-ray crystallography, the structure is often accompanied by crystallographic water molecules, and could also contains substrates or ions. We will then use biopython to clean up the PDB structure from non-protein atoms."]},{"cell_type":"markdown","metadata":{"id":"JKUTMHeklYdC"},"source":["1. Download the structure of *T. hennahi* myohemerythrin (PDB 2mhr) using biopython:"]},{"cell_type":"code","metadata":{"id":"Xd7FJOtplOF0"},"source":["#Importing your PDB file using biopython\n","import os\n","from Bio.PDB import *\n","\n","#Here, we add a unique or multiple PDB accession IDs\n","pdbid = ['2mhr']\n","\n","#We will treat the IDs as a list to download all PDBs\n","pdbl = PDBList()\n","for s in pdbid:\n","  pdbl.retrieve_pdb_file(s, pdir='.', file_format =\"pdb\", overwrite=True)\n","  os.rename(\"pdb\"+s+\".ent\", s+\".pdb\")\n","print(\"DONE!\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ZRorFpoyzV1p"},"source":["2. Clean up the protein structure from crystallographic waters and other non-protein atoms with biopython:"]},{"cell_type":"code","metadata":{"id":"Nj2j6TH9eUvt"},"source":["#Here we set up a parser for our PDB\n","parser = PDBParser()\n","io=PDBIO()\n","structure = parser.get_structure('A', '2mhr.pdb')\n","#Now we remove hydrogens, waters and ligands using Dice\n","#and save the cleaned-up structure with a different filename\n","io.set_structure(structure)\n","sel = Dice.ChainSelector('A', 1, 118)\n","io.save(\"2mhr_A.pdb\", sel)\n","print(\"Your PDB was processed. Only the protein heavy atoms have been kept\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gHQFin-HYxRt"},"source":["3. Lastly, we will visualize the three-dimensional structure of *T. hennahi* myohemerythrin using py3Dmol.\n","\n","**NOTE‚ùóÔ∏è** For other color schemes, check the available ones at the [3Dmol](https://3dmol.csb.pitt.edu/doc/types.html#ColorschemeSpec) website."]},{"cell_type":"code","metadata":{"id":"-oGLjJmdYh1-"},"source":["import py3Dmol\n","#First we assign the py3Dmol.view as view\n","view=py3Dmol.view()\n","#The following lines are used to add the addModel class\n","#to read the PDB files\n","view.addModel(open('2mhr_A.pdb', 'r').read(),'pdb')\n","#Here we set the background color as white\n","view.setBackgroundColor('white')\n","#Here we set the visualization style and color\n","view.setStyle({'chain':'A'},{'cartoon': {'colorscheme':'ssJmol'}})\n","#You can activate the labels for each residue if you want\n","#or comment them with a '#' at the beggining of each line if you do not want to\n","view.addResLabels({'resi':'1'},{'fontColor':'white','fontOpacity':1,'showBackground':'true'})\n","view.addResLabels({'resi':'118'},{'fontColor':'white','fontOpacity':1,'showBackground':'true'})\n","\n","#Here we center the molecule for its visualization\n","view.zoomTo()\n","#And we finally visualize the structures using the command below\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Z65j3jScfS18"},"source":["**QUESTIONS‚ùì**\n","- How many helices does this structure have?\n","- How is the topological order of the secondary structure elements in the three-dimensional structure?\n","- Is there a protein region that we could intuitively identify as disordered?"]},{"cell_type":"markdown","metadata":{"id":"ZJas1A0hffk-"},"source":["#Part II ‚Äì Generate a multiple sequence alignment (MSA) using HMMER "]},{"cell_type":"markdown","metadata":{"id":"p77vPJ1lgLEh"},"source":["As illustrated in FIG 1, a multiple sequence alignment (MSA) of many sequences of a protein family is required to infer physical interactions. A natural question that arises is where to obtain such sequences.\n","\n","As previously indicated, *T. hennahi* myohemerythrin is a member of the **Hemerythrin family** in the **Pfam** database [**PF01814**](https://pfam.xfam.org/family/PF01814). The benefit of Pfam is that it contains **seed alignments**, i.e. small MSA based on a subset of known sequences that are used to **construct profile HMMs** to search for other protein homologs or to generate a larger MSA, as well as **readily available MSAs of identified sequences using these HMM profiles**.\n","\n","We will use these seed alignments and known Hemerythrin family protein sequences as inputs for generating a HMM-based MSA using HMMER."]},{"cell_type":"markdown","metadata":{"id":"7CDSHg9JeABH"},"source":["1. We will first create a folder (**_dca_analysis_**) in which we will save all of the input and output files from our coevolutionary analysis."]},{"cell_type":"code","metadata":{"id":"7lMyxnTmtmwC"},"source":["#Let's make a folder for the coevolutionary analysis\n","#We need to import the os and path library\n","import os\n","from pathlib import Path \n","\n","#Then, we define the path of the folder we want to create.\n","#Notice that the HOME folder for a hosted runtime in colab is /content/\n","dcapath = Path(\"/content/dca_analysis/\")\n","\n","#Now, we create the folder using the os.mkdir() command\n","#The 'if' conditional is just to check whether the folder already exists\n","#In which case, python returns an error\n","if os.path.exists(dcapath):\n","  print(\"path already exists\")\n","if not os.path.exists(dcapath):\n","  os.mkdir(dcapath)\n","  print(\"path was succesfully created\")\n","\n","#We finally move onto this new folder\n","os.chdir(dcapath)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Rm4tR7SBueva"},"source":["2. We will now download the sequence of *T. hennahi* myohemerythrin (PDB 2mhr) in FASTA format using biopython.\n","\n","**NOTE‚ùóÔ∏è** Given that PDB files can contain more than one polypeptide chain, it is required to specify the chain ID from which the protein sequence will be extracted. In our case, 2MHR only contains a single chain A, thus our accession ID is *2MHR_A*"]},{"cell_type":"code","metadata":{"id":"XXQ19ku5uipd"},"source":["import os\n","from pathlib import Path \n","from Bio import SeqIO, Entrez\n","from Bio.SeqRecord import SeqRecord\n","#Indicate the accession ID of the protein sequence to download.\n","seqlist = ['2MHR_A']\n","for n in seqlist:\n","  #Setting up your email to be able to use Entrez\n","  Entrez.email = 'your.email@mail.com'\n","  #Here, we set up a temporary handle with our downloaded sequence\n","  #in fasta format\n","  temp = Entrez.efetch(db=\"sequences\",rettype=\"fasta\",id=n)\n","  #Creating a fasta file to write our downloaded sequence\n","  seq_out = open(\"2mhr_A.fasta\",'w')\n","  #Reading the sequence information as a string in fasta format\n","  seq = SeqIO.read(temp, format=\"fasta\")\n","  print(\"The downloaded protein sequence is:\\n\" + seq.seq + \"\\n\")\n","  SeqIO.write(seq,seq_out,\"fasta\")\n","  #Closing both the temp handle and the FASTA file\n","  temp.close()\n","  seq_out.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"oO5pJY7kPfss"},"source":["3. Then, we will go to the [**Pfam** database](https://pfam.xfam.org) and search for the appropriate protein family. You can easily do this by inputting **_Hemerythrin_** or **_PF01814_** in the **_Jump To_** search box. Once here, we will obtain two different files. The first one is the **seed alignment**.\n","\n","- Go to the **_Alignment_ ‚Üí Format an alignment** menu and **download the seed alignment**, which is used for constructing a profile HMM. Download this seed alignment in either STOCKHOLM or FASTA format.\n","\n","  Here, we simplify this exercise by directly downloading the seed alignment in Google Colab as a text file (*PF01814_seed*) and using `hmmbuild` from HMMER to generate our profile HMM."]},{"cell_type":"code","metadata":{"id":"g6M_sPWk5icD"},"source":["!wget -O PF01814_seed https://pfam.xfam.org/family/PF01814/alignment/seed\n","!hmmbuild PF01814.cm PF01814_seed"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"fHI-ahHvjGLt"},"source":["- The second file is **a set of unaligned sequences from one of the available databases** in FASTA format, making sure to select the **`no gaps (unaligned)`** option in the **GAPS** section. We will add our target sequence to this sequence set (which is why we are using a FASTA format) and generate a MSA using our generated profile HMM and HMMER.\n","  \n","  While the higher the number of sequences the better, we **highly recommend to download the sequences from RP15, RP35 or RP55** due to our time constraints. \n","\n","  Once downloaded, import this file onto Google Colab by dragging it into the appropriate folder in the **Files** tab, located on the left sidebar of Google Colab, and rename it as **_PF01814_seqs.txt_**"]},{"cell_type":"markdown","metadata":{"id":"qqP6A3iAANbh"},"source":["**‚ö†Ô∏èWARNING:** We have recently checked that the PFam website is a little bit slow. Hence, we have uploaded the sequences for PF01814 onto our GitHub. Please choose between RP15, RP35, RP55, RP75 or UNIPROT as shown below:"]},{"cell_type":"code","metadata":{"id":"RYwZGtYAAmn1"},"source":["#You can select between RP15, RP35, RP55, RP75 or UNIPROT by changing the suffix\n","#Use lowercase!\n","%%bash\n","suffix=rp55\n","wget -O PF01814_seqs.txt https://github.com/pb3lab/ibm3202/raw/master/files/pfam_files/PF01814_$suffix.txt"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"lias6SkMSdSL"},"source":["4. Then, we will add our target sequence to our sequence set and use the profile HMM to generate a MSA using `hmmalign` from HMMER. The resulting MSA will be stored as **_PF01814_2mhr_aligned.afa_**"]},{"cell_type":"code","metadata":{"id":"2LoavSBKb9N2"},"source":["!cat PF01814_seqs.txt 2mhr_A.fasta > PF01814_2mhr.fasta\n","!hmmalign -o PF01814_2mhr_aligned.afa --outformat AFA PF01814.cm PF01814_2mhr.fasta"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"pmx4F5DeU-fS"},"source":["#Part III ‚Äì Infer physical interactions between residue pairs in an MSA through coevolutionary analysis on pyDCA"]},{"cell_type":"markdown","metadata":{"id":"sdsh4ZSZVKVe"},"source":["The resulting MSA will be our input for our coevolutionary analysis using the **mean-field approximation of DCA (mfDCA)**, one of the two available approaches (the other being pseudo-maximum likelihood or plmDCA) in pyDCA, with mfDCA being faster to compute. \n","\n","pyDCA is a python implementation of DCA that not only infers physical contacts from the MSA, but also ascertains the accuracy of our prediction by establishing the number of true positives when compared to the contacts observed in a solved protein structure.\n","\n","Briefly, DCA will allow for disentangling direct contributions to correlations (resulting from native contacts) from indirect contributions (mediated through chains of native contacts). Given that our intention is to predict the structure of *T. hennahi* myohemerythrin based on sequence information alone, we will use its protein sequence as reference for the prediction of residue-pair contacts.\n","\n","**NOTE‚ùóÔ∏è** While the statistical physics algorithms available in pyDCA are outside ths scope of this tutorial, a deep dive into mfDCA and plmDCA can be found in:\n","- Ekeberg, M., L√∂vkvist, C., Lan, Y., Weigt, M., & Aurell, E. (2013). Improved contact prediction in proteins: using pseudolikelihoods to infer Potts models. Physical Review E, 87(1), 012707."]},{"cell_type":"markdown","metadata":{"id":"Qz2lSrMbl9m5"},"source":["1. We will first import all of the required modules for pyDCA"]},{"cell_type":"code","metadata":{"id":"RKZyDUE7vayJ"},"source":["from pydca.plmdca import plmdca\n","from pydca.meanfield_dca import meanfield_dca\n","from pydca.sequence_backmapper import sequence_backmapper\n","from pydca.msa_trimmer import msa_trimmer\n","from pydca.contact_visualizer import contact_visualizer\n","from pydca.dca_utilities import dca_utilities"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"q-4JkIS9WMnM"},"source":["2. Then, we will trim our MSA file based on the lenght of the target sequence, which in this case corresponds to the sequence of *T. hennahi* myohemerythrin that was solved by X-ray crystallography. The input MSA is defined in the `prot_msa_file` path, whereas our target sequence is defined in the `prot_refseq_file`. The trimmed MSA data to an output file (**_MSA_PF01814_2mhr_trimmed.fa_**) in FASTA format."]},{"cell_type":"code","metadata":{"id":"I5VPLbMoWFdS"},"source":["#Paths to our MSA and reference sequence files\n","\n","prot_msa_file = 'PF01814_2mhr_aligned.afa'\n","prot_refseq_file = '2mhr_A.fasta'\n","\n","#Create MSATrimmer instance \n","trimmer = msa_trimmer.MSATrimmer(\n","    prot_msa_file, biomolecule='protein', \n","    refseq_file=prot_refseq_file,\n",")\n","\n","trimmed_data = trimmer.get_msa_trimmed_by_refseq(remove_all_gaps=True)\n","\n","#write trimmed MSA to file in FASTA format\n","trimmed_data_outfile = 'MSA_PF01814_2mhr_trimmed.fa'\n","with open(trimmed_data_outfile, 'w') as fh:\n","    for seqid, seq in trimmed_data:\n","        fh.write('>{}\\n{}\\n'.format(seqid, seq))"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"7leSC5tIXJ6Y"},"source":["3. Once done, we will compute our DCA scores using the **mean-field approach (mfDCA)**. For this analysis, we first create a mfDCA instance `mfdca_inst` to analyze our trimmed MSA. We also provide a series of optional parameters:\n","- A pseudocount, which is basically adding extra ‚Äúpseudo‚Äù observations to the real data in order to cure singularities caused by strong correlations in undersampled data, using the `pseudocount` parameter. This value is typically set to one-third (0.3) to one-half (0.5) of the total numer of sequences in the MSA. \n","- A maximum sequence identity percentage for all sequences in the MSA, using the `seqid` parameter. This value is typically set to 80% (0.8) sequence identity.\n","- The number of threads, which is set to 2, the maximum for Google Colab.\n","\n","  After the fields and couplings are obtained, the DCA scores are computed from the Frobenius norm of the couplings between sites $i$ and $j$ by calling the `compute_sorted_FN_APC()` method on `mfdca_inst`. This action returns the average product corrected (APC) DCA scores.\n"]},{"cell_type":"code","metadata":{"id":"PS1M7IVNs_UE"},"source":["# Compute DCA scores using the mean-field algorithm\n","mfdca_inst = meanfield_dca.MeanFieldDCA(\n","    trimmed_data_outfile,\n","    'protein',\n","    pseudocount = 0.5,\n","    seqid = 0.8,\n",")\n","\n","# Compute average product corrected Frobenius norm of the couplings\n","mfdca_FN_APC = mfdca_inst.compute_sorted_FN_APC()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"mhWuPHSggJNY"},"source":["4. Once these mfDCA calculations are finished, we can print them out on screen. Here, we will use python to print the top 10 DCA pairs."]},{"cell_type":"code","metadata":{"id":"MGRB2esLtmHo"},"source":["for site_pair, score in mfdca_FN_APC[:10]:\n","    print(site_pair, score)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"LC9czoh-gPcL"},"source":["5. A more meaningful way to determine how good are our predictions of physical contacts is to visually and statistically compare them against experimental data.\n","\n","  For this purpose, we will generate a `DCAVisualizer` to compare our inferred contact map against the experimental one derived from the protein structure of *T. hennahi* myohemerythrin.\n","  \n","  This visualizer takes the type of biomolecule (`'protein'`), our reference sequence file (`refseq_file`) and our previously generated list of sorted DCA scores (`sorted_dca_scores`) as input to generate a contact map based on sequence information alone.\n","\n","  When structural information is available, it also takes a PDB ID (`'2mhr_A'`) and chain (`'A'`) as input parameters, as well as specified linear (i.e. sequence separation between residue pairs, `linear_dist`) and contact distances (`contact_dist`, in √Ö)."]},{"cell_type":"code","metadata":{"id":"4kyXoQnbtsZ-"},"source":["mfdca_visualizer = contact_visualizer.DCAVisualizer('protein', 'A', '../2mhr_A.pdb',\n","    refseq_file = prot_refseq_file,\n","    sorted_dca_scores = mfdca_FN_APC,\n","    linear_dist = 4,\n","    contact_dist = 8.0,\n",")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ja5UOuOgiOe-"},"source":["6. We will now plot both contact maps, with the structure-based contacts on the upper left triangle, and the DCA-based contacts on the lower right triangle. Correctly predicted contacts are shown in green, whereas false positives are shown in red. Also note that the number of DCA contacts is equivalent to $L$, where $L$ is the lenght (i.e. number of columns) of the trimmed alignment."]},{"cell_type":"code","metadata":{"id":"-EVO_A29t1FF"},"source":["contact_map_data = mfdca_visualizer.plot_contact_map()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"qusYWvKOiulO"},"source":["7. Lastly, we will determine the accuracy of our coevolutionary analysis by looking at its true positive (TP) rate per rank. The TP rate per rank is the number of correctly predicted contacts per rank of the predicted pairs divided by all predictions at that rank.\n","\n","  In the resulting plot, the blue line will correspond to the TP rates for the predicted contacts, and the orange line to the theoretically maximum possible true positive rate for the contacts obtained from the experimental structure."]},{"cell_type":"code","metadata":{"id":"POcX6okSt7GA"},"source":["tp_rate_data = mfdca_visualizer.plot_true_positive_rates()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"tBZAiV-ffXbe"},"source":["**QUESTIONS‚ùì**\n","- How would the number of true positive contacts change as a function of the number of sequences in the MSA?\n","- How would the number of true positive contacts change as a function of the sequence identity in the MSA?\n"]},{"cell_type":"markdown","metadata":{"id":"GEHOpdBWf5eT"},"source":["#Part IV ‚Äì Generate an SBM based on DCA and secondary structure information for protein structure prediction"]},{"cell_type":"markdown","metadata":{"id":"eLYRWh4vgfhq"},"source":["SBM are based on the treatment of **native contacts**, i.e. pairs of residues that fall below a distance cutoff in a given protein structure, as **attractive non-bonded interactions through either Lennard-Jones or Gaussian potentials**, whereas all other non-native interactions are treated as repulsive.\n","\n","In this context, **the use of physical interactions between residue pairs inferred by DCA** instead of experimentally defined native contacts is a natural extension to SBM, due to the simplicity of its attractive component in the energy function.\n","\n","Here, we will use our predicted contacts, along with bonded potentials (bonds, angles and dihedrals) drawn based on secondary structure predictions using **Jpred4**, to generate an SBM that enables protein folding simulations to predict the structure of *T. hennahi* myohemerythrin based on sequence information alone.\n","\n","**NOTE‚ùóÔ∏è** While the details of sequence-based secondary structure prediction are outside the scope of this tutorial, a good primer is found in the following publication:\n","\n","- Rost, B. (2001). Protein secondary structure prediction continues to rise. Journal of structural biology, 134(2-3), 204-218."]},{"cell_type":"markdown","metadata":{"id":"n0Jr-t8akWm5"},"source":["1. We will start by creating a new folder in which we will generate our SBM for *T. hennahi* myohemerythrin"]},{"cell_type":"code","metadata":{"id":"scIhDT82u91k"},"source":["#Let's make a folder for creating our structure-based model (SBM)\n","sbmpath = Path(\"/content/sbm/\")\n","\n","if os.path.exists(sbmpath):\n","  print(\"path already exists\")\n","if not os.path.exists(sbmpath):\n","  os.mkdir(sbmpath)\n","  print(\"path was succesfully created\")\n","\n","#Changing directory to the sbm folder\n","os.chdir(sbmpath)\n","\n","#We will also copy the cleaned up PDB file of 2MHR\n","!cp ../2mhr_A.pdb experimental.pdb"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"z7T4yxpMkeZu"},"source":["2. Now, we will store the top $L$ DCA contacts, where $L$ is the length of the (i.e. number of columns) of the MSA, as a 3-column text file with the residue pairs in columns 1 and 2 and the DCA scores in column 3.\n","\n","  These contacts will be used to define **non-bonded attractive interactions betweeen residue pairs** in our SBM."]},{"cell_type":"code","metadata":{"id":"DCTZXKMr6BYf"},"source":["#Set the L length, equivalent to 'Number of DCA contacts' in our previous plots\n","Llength = 118\n","#Saving the top L DCA contacts with i > j+3\n","predicted_contact_map = 'predicted_contact_map.txt'\n","\n","#Writing top L DCA contacts\n","with open(predicted_contact_map, 'w') as cm:\n","  count=0\n","  for site_pair, score in mfdca_FN_APC:\n","    if site_pair[1]>site_pair[0]+3:\n","      count += 1\n","#Note that we have to add 1 to each residue i,j\n","      cm.write('{0} {1} {2}\\n'.format(site_pair[0]+1, site_pair[1]+1, score))\n","    if count == Llength:\n","      break"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"DEH5XmC5loKz"},"source":["3. We will also need a secondary struture prediction to generate the **bonded interactions** of our SBM. For this, we will visit the [Jpred4 webserver](http://www.compbio.dundee.ac.uk/jpred/index.html), input our sequence in the text box and click on **`Make a Prediction`**, confirm that you still want to make a prediction by clicking on **`Continue`** and then select to **`View simple results in HTML`**.\n","\n","  Then, load the first and second lines of the results obtained by Jpred4 in the following text box.\n"]},{"cell_type":"code","metadata":{"id":"noA2HIMCy5q1","cellView":"form"},"source":["#@title Enter the amino acid sequence and predicted secondary structure of your protein\n","sequence = ''  #@param {type:\"string\"}\n","Jpred = ''  #@param {type:\"string\"}\n","f = open(\"Jpred.txt\", \"a\")\n","f.write(sequence+\"\\n\")\n","f.write(Jpred)\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"EwJfImwJnaA6"},"source":["4. The tertiary contacts inferred by DCA and the secondary structure predicted by Jpred4 will now be used to generate our SBM using the script below. The result from this script are:\n","- A **coordinate** file (extension **.gro**), where each residue of the target sequence is represented by a single bead and given initial positions for starting our folding simulations.\n","- A **topology** file (extension **.top**), which contains the non-bonded interactions from the DCA contacts as Gaussian potentials and the bonded interactions (bonds, angles, dihedrals) drawn from the secondary structure prediction as harmonic potentials. \n","\n","  Upon executing this script, you will be asked to indicate a `DCA maximum force factor`, which controls the force of pairwise potentials that will drive protein folding. A good estimate for the strength of coevolutionary interactions is to consider the force factor equivalent to the ratio of $L$ over the number of DCA pairs (factor = $L$/$|$DCA$|$). Thus, **we will use a value of 1.**"]},{"cell_type":"code","metadata":{"id":"FVOwnB9VuwEv"},"source":["#Downloading all required files\n","!wget https://github.com/pb3lab/ibm3202/raw/master/scripts/dcasbm.py\n","!wget https://github.com/pb3lab/ibm3202/raw/master/scripts/distavg\n","!wget https://github.com/pb3lab/ibm3202/raw/master/files/sbm_calpha_SA_v5_short.mdp"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"JzDBFKLC0gNo"},"source":["!python dcasbm.py Jpred.txt predicted_contact_map.txt"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"lUNHELyRp5J0"},"source":["5. Lastly, we will use our coordinate and topology files along with a parameter file to set up and run a folding simulation in our SBM-enhanced version of GROMACS.\n","\n","  These simulations start at a high temperature in which the protein is unfolded, and then the system temperature is gradually reduced until the protein reaches an energy minimum where most interactions are satisfied."]},{"cell_type":"markdown","metadata":{"id":"23pzmGVIkPQ3"},"source":["**‚ö†Ô∏èWARNING:** There is an unresolved issue with the use of Gaussian nonbonded potentials on the SBM-enhanced version of GROMACS over Google Colab, which causes errors in the simulations when running for long timesteps. Typically, this will be reflected on the run.gro file by showing **`nan`** instead of coordinate positions for all atoms. Thus, we limited ourselves to very short simulations in this tutorial. If you have this error, try to repeat your simulations. **Please proceed with caution!**"]},{"cell_type":"code","metadata":{"id":"kojZfDKD4ZNS"},"source":["%%time\n","%%bash\n","source /content/gromacs_sbm/bin/GMXRC\n","#Preparing the binary input for our folding simulation file\n","gmx grompp -f sbm_calpha_SA_v5_short.mdp -c Jpred_calpha.gro -p Jpred_calpha.top -o run.tpr\n","#Running our folding simulation\n","gmx mdrun -deffnm run -nt 1 -noddcheck"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"JgIiiRTnskQd"},"source":["üò± **EMERGENCY BACKUP!** Your simulation failed?! Use the following code cell to download a successful simulation run:"]},{"cell_type":"code","metadata":{"id":"DzeOqhS4sjcU"},"source":["#Use only in case of emergency\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/files/emergency_backup/lab13/dca_sbm_backup.tar.gz\n","!tar xzf dca_sbm_backup.tar.gz"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"TL7UIxuEtEbx"},"source":["6. We will now convert our final .gro structure into a PDB file for further comparison with the experimental structure of *T. hennahi* myohemerythrin "]},{"cell_type":"code","metadata":{"id":"R4L4JBfM2Eax"},"source":["%%bash\n","source /content/gromacs_sbm/bin/GMXRC\n","#Converting our final structure from .gro into a PDB file\n","gmx editconf -f run.gro -o predicted.pdb"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"pkQOavMWtOu5"},"source":["7. Once our folding simulation is complete and we have converted our final coordinates onto the PDB file format, we can compare the predicted structure based on sequence information alone with the experimentally determined structure of *T. hennahi* myohemerythrin. For this, we will first perform a structural alignment of both structures.\n","\n","  In the following box, please enter the filename of the experimental structure (including its .pdb extension) and indicate the starting and ending residues that you wish to align (e.g. 1 to 118)\n","  \n","**NOTE‚ùóÔ∏è** Remember that, typically, the N- and C-termini of a protein structure are very flexible and can be disregarded from the structural alignmen to obtain better results."]},{"cell_type":"code","metadata":{"id":"GpJplQzgHaQl","cellView":"form"},"source":["#@title Perform a structural alignment between the predicted and experimental structure\n","\n","import Bio.PDB\n","\n","# Select what residues numbers you wish to align\n","# and put them in a list\n","\n","#@markdown 1. Indicate the starting and ending residues you wish to align \n","\n","start_resi = 17 #@param {type:\"number\"}\n","end_resi   = 118 #@param {type:\"number\"}\n","atoms_to_be_aligned = range(start_resi, end_resi + 1)\n","\n","# Start the parser\n","pdb_parser = Bio.PDB.PDBParser(QUIET = True)\n","\n","#@markdown 2. Indicate the filename of the experimental structure (PDB file) \n","\n","# Get the structures\n","experimental = 'experimental.pdb' #@param {type:\"string\"}\n","predicted = 'predicted.pdb'\n","ref_structure = pdb_parser.get_structure(\"reference\", experimental)\n","sample_structure = pdb_parser.get_structure(\"sample\", predicted)\n","\n","# Use the first model in the pdb-files for alignment\n","# Change the number 0 if you want to align to another structure\n","ref_model    = ref_structure[0]\n","sample_model = sample_structure[0]\n","\n","# Make a list of the atoms (in the structures) you wish to align.\n","# In this case we use CA atoms whose index is in the specified range\n","ref_atoms = []\n","sample_atoms = []\n","\n","# Iterate of all chains in the model in order to find all residues\n","for ref_chain in ref_model:\n","  # Iterate of all residues in each model in order to find proper atoms\n","  for ref_res in ref_chain:\n","    # Check if residue number ( .get_id() ) is in the list\n","    if ref_res.get_id()[1] in atoms_to_be_aligned:\n","      # Append CA atom to list\n","      ref_atoms.append(ref_res['CA'])\n","\n","# Do the same for the sample structure\n","for sample_chain in sample_model:\n","  for sample_res in sample_chain:\n","    if sample_res.get_id()[1] in atoms_to_be_aligned:\n","      sample_atoms.append(sample_res['CA'])\n","\n","# Now we initiate the superimposer:\n","super_imposer = Bio.PDB.Superimposer()\n","super_imposer.set_atoms(ref_atoms, sample_atoms)\n","super_imposer.apply(sample_model.get_atoms())\n","\n","# Print RMSD:\n","print('The calculated RMSD is:')\n","print (super_imposer.rms)\n","\n","# Save the aligned version of one of the chains of 6ANE\n","io = Bio.PDB.PDBIO()\n","io.set_structure(sample_structure) \n","io.save(\"predicted_aligned.pdb\")\n","\n","aligned = 'predicted_aligned.pdb'"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"SUr08gPrOd3Z"},"source":["7. Now we can load the superimposed structure into py3dmol with the code cell shown below to visualize the accuracy of our prediction.\n","\n","**QUESTION‚ùì**\n","- How accurate was your prediction when compared to the experimentally solved structure of *T. hennahi* myohemerythrin?\n","- In which ways do you believe that the accuracy of your prediction can be improved?"]},{"cell_type":"code","metadata":{"id":"jWHfH9RyJR0D"},"source":["view=py3Dmol.view()\n","view.addModel(open(experimental, 'r').read(),'pdb')\n","view.setStyle({'chain':'A'},{'cartoon': {'color':'spectrum'}})\n","view.addModel(open(aligned, 'r').read(),'pdb')\n","view.setStyle({'chain':' '},{'cartoon': {'style':'trace','color':'red'}})\n","#You can opt to delete these labels\n","view.addResLabels({'resi':'1'},{'fontColor':'white','fontOpacity':1,'showBackground':'true'})\n","view.addResLabels({'resi':'118'},{'fontColor':'white','fontOpacity':1,'showBackground':'true'})\n","\n","view.zoomTo()\n","view.setBackgroundColor('white')\n","view.show()"],"execution_count":null,"outputs":[]}]}