{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"-IGnGd85EHFG"},"source":["#Lab.04 / IBM3202 ‚Äì¬†Comparative Modeling using MODELLER"]},{"cell_type":"markdown","metadata":{"id":"RVDb8Wr5EJeT"},"source":["##Theoretical Aspects"]},{"cell_type":"markdown","metadata":{"id":"bWG8g943ENJP"},"source":["As we discussed during our Lectures, **the structure space is rather small when compared with the sequence space**. In fact, hierarchical structure classifications such as **CATH** have demonstrated that only a few proteins of the enormous pool of structures being deposited each year in the Protein Data Bank are known to provide novel folds. In this regard, it seems that we have discovered almost all protein folds. Moreover, the elegant experiment by Chothia & Lesk [Chothia C & Lesk AM (1986) _EMBO J_ 5(4), 823‚Äì826] revealed that, generally, **protein structures are quite similar between proteins when their sequence identity is > 30-40%**, with some remarkable exceptions to this rule)."]},{"cell_type":"markdown","metadata":{"id":"zGAwxMw3H3mZ"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/cm_01.png'/>\n","<figcaption>FIGURE 1. Relationship between sequence identity and structural similarity for globular and membrane proteins.\n","<br>Olivella M et al (2013) <i>Bioinformatics 29(13), 1589-1592</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"K8lKUSvLWkVQ"},"source":["This evidence provides the perfect framework to model the structure of a given protein based on another protein (a template) for which its structure has been solved and that has sufficient sequence identity. But how does it work? Well, basically it operates by using a lot of distance constraints.\n","\n","Similarly to what we saw for coevolutionary signals in our Lectures, in which pairs of residues are conserved regardless of their sequence separation because of their role in stabilizing the native state through interactions in the 3D space, we can add interaction constraints to a sequence of a protein with unknown structures based on what we observe on the template structure."]},{"cell_type":"markdown","metadata":{"id":"eB0BhtM7H7al"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/cm_02.png'/>\n","<figcaption>FIGURE 2. From sequence similarity to spatial constraints for comparative modelling in MODELLER.\n","<br>Fiser A & Sali A (2003) <i>Methods Enzymol 374, 461-491</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"s2CcC8JQETHX"},"source":["##Overview"]},{"cell_type":"markdown","metadata":{"id":"Is36qLm3EWcp"},"source":["in this tutorial, we will pursue the template-based modeling of an enzyme homologous to _I. sakaiensis_ PET hydrolase (PETase) [Fecker T et al (2018) *Biophys J 114 (6), 1302-1312*] on the software **MODELLER** using a single-template approach (i.e. based on only one protein structure). Then, we will analyze the model and determine its quality using an online server."]},{"cell_type":"markdown","metadata":{"id":"xMf4RZ9mRGoP"},"source":["#Part 0 ‚Äì Downloading and Installing the required software"]},{"cell_type":"markdown","metadata":{"id":"o2o6UeVMUNBg"},"source":["Before we start, you must first **remember to start the hosted runtime in Google Colab**.\n","\n","Then, we must install several pieces of software to perform this tutorial. Namely:\n","- **MODELLER**, a famous software for template-based modelling.\n","- **py3Dmol**, for visualization of the template and modelled protein structures.\n","- **biopython**, for manipulation and retrieval of protein structures and sequences.\n","\n","After several tests, the following installation instructions are the best way of setting up **Google Colab** for this laboratory session.\n","\n","1. We will first install MODELLER as follows:\n","\n","‚ö†Ô∏è**WARNING!:** In order to use MODELLER, you will need to obtain an Academic License by registering **[in this website](https://salilab.org/modeller/registration.html)**. The license key will be immediately sent to your email address."]},{"cell_type":"code","metadata":{"id":"Dct8-bXnbwyD"},"source":["#Before running this script, make sure to replace the MODELLER\n","#License Key with the one sent after registration in the MODELLER website\n","!wget https://salilab.org/modeller/10.4/modeller-10.4.tar.gz\n","#Then, we extract the downloaded folder containing MODELLER 10.1\n","!tar -zxf modeller-10.4.tar.gz\n","!echo \"MODELLER extraction completed\"\n","#Then, we switch onto the MODELLER folder\n","#with an automagic command\n","%cd modeller-10.4\n","#And we prepare a file containing the minimal setup elements\n","#For installing, including a license key\n","with open('modeller_config', 'a') as f:\n","  f.write(\"2\\n\")\n","  f.write(\"/content/compiled/MODELLER\\n\")\n","#ADD YOUR LICENSE KEY HERE!\n","  f.write(\"MODELIRANJE\\n\")\n","!./Install < modeller_config\n","!echo \"MODELLER set up completed\"\n","%cd /content/"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"hc7elmkrkRsi"},"source":["#Creating a symbolic link\n","%cd modeller-10.4\n","!ln -sf /content/compiled/MODELLER/bin/mod10.4 /usr/bin/\n","%cd /content/\n","#Checking if MODELLER works\n","!mod10.4 | awk 'NR==1{if($1==\"usage:\") print \"MODELLER succesfully installed\"; else if($1!=\"usage:\") print \"Something went wrong. Please install again\"}'"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"iqxGqQiBfH01"},"source":["2. Then, we will install biopython"]},{"cell_type":"code","metadata":{"id":"PGNoNof2RQFh"},"source":["#Installing biopython using pip\n","!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"T7V_RsmMXViv"},"source":["3. Finally, we will install py3Dmol"]},{"cell_type":"code","metadata":{"id":"fGwT3ZARyklF"},"source":["#Installing py3Dmol using pip\n","!pip install py3Dmol\n","#And importing the py3Dmol module\n","import py3Dmol"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2-3uHREEEaKG"},"source":["#Part\tI ‚Äì Retrieving the amino acid sequence of PETase homologs"]},{"cell_type":"markdown","metadata":{"id":"0uulsIsuEgEA"},"source":["Typically, this step starts with a **BLAST search** (or other types of database search) for sequence homologs of a given protein of interests, taking into consideration the e-values, sequence coverages and identities.\n","\n","Here, we skipped the BLAST search that we performed during our previous tutorial and will just download one sequence.\n","\n","1. Download the sequence with accession ID BAB86909.1 using biopython as we did in previous tutorials:"]},{"cell_type":"code","metadata":{"id":"M5TsXqgwfZ6U"},"source":["import os\n","from pathlib import Path\n","from Bio import SeqIO, Entrez\n","seqlist = ['6ANE_A', 'BAB86909.1']\n","for n in seqlist:\n","  #Creating folder for our sequence\n","  if not os.path.exists(n):\n","    os.mkdir(n)\n","  folder = Path(n)\n","  #Setting up your email to be able to use Entrez\n","  Entrez.email = 'your.email@uc.cl'\n","  #Here, we set up a temporary handle with our downloaded sequence in fasta format\n","  temp = Entrez.efetch(db=\"protein\",rettype=\"fasta\",id=n)\n","  #Creating a fasta file to write our downloaded sequence\n","  aaseq_out = open(folder/ \"target.fasta\",'w')\n","  #Reading the sequence information as a string in fasta format\n","  aaseq = SeqIO.read(temp, format=\"fasta\")\n","  #Writing the sequence record in fasta format\n","  SeqIO.write(aaseq,aaseq_out,\"fasta\")\n","  #Closing both the temp handle and the FASTA file\n","  temp.close()\n","  aaseq_out.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"R0__pRKSEoAt"},"source":["2. If everything worked smoothly, your sequence should have been downloaded as a FASTA file in an appropriate folder named after its accession ID. Could you check what is the description and sequence of this file?\n","\n","**üí° HINT:** You will only retrieve the information for the last query in the list, as you are overwriting *aaseq*"]},{"cell_type":"code","metadata":{"id":"jxtqEw7eiKRm"},"source":["#Obtain the description of the downloaded protein\n","print(aaseq.description)\n","#Obtain the sequence of the downloaded protein\n","print(aaseq.seq)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"DdjTrTTEZrwb"},"source":["3. It is highly recommended at this point that you change the name of the sequence (labeled as ‚Äú>‚Äù) to something shorter, such as just the ID or a name (e.g. target). You can simply do this manually by opening this file on Google Colab. **For this tutorial, we are indeed changing the name to target**"]},{"cell_type":"markdown","metadata":{"id":"CJHr933JIu0j"},"source":["#Part\tII ‚Äì Select an appropriate template structure and perform a sequence alignment for protein structure modeling"]},{"cell_type":"markdown","metadata":{"id":"TF9HsgUtEuAv"},"source":["Selecting an appropriate template for modeling a structure of a homologous protein is as crucial as an appropriate alignment to correctly position the different residues.\n","\n","**QUESTION:** What features of a crystal structure do you think are important for choosing the best template?\n","\n","1. As we will  work on the sequence BAB86909.1, we will first change directory."]},{"cell_type":"code","metadata":{"id":"IVEC8dLlScXG"},"source":["%cd BAB86909.1"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Fji-N_jhSgQJ"},"source":["2. We will find the potential best templates from the whole PDB database to model the structure of our target protein using the **profile.build()** command from MODELLER.\n","\n","  For this purpose, we need a text file containing a list of non-redundant PDB sequences at 95% sequence identity and an appropriate script for running MODELLER."]},{"cell_type":"code","metadata":{"id":"-XvmVrRdRrMm"},"source":["#Downloading pdb_95.pir\n","!wget https://salilab.org/modeller/downloads/pdb_95.pir.gz\n","!gunzip pdb_95.pir.gz\n","#Downloading the build_profile.py script from GitHub\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/scripts/build_profile.py"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"2bZ0bgu6SkOr"},"source":["#Running the build_profile script\n","!mod10.4 build_profile.py\n","#Printing only the list of potential templates\n","!sed -n '/HITS FOUND IN ITERATION:     1/,/Weight Matrix/p;/Weight Matrix/q' build_profile.log"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ekC7m-xcQ72Q"},"source":["In this particular example, a BLOSUM62 similarity matrix is being used for determining the sequence identity between target and potential templates. Also, we are employing only one search iteration and the parameter max_aln_evalue is set to 0.01, indicating that only sequences with e-values smaller than or equal to 0.01 will be included in the final profile.\n","\n","For simplicity, we just printed out the PDB table from the resulting log file generated during this analysis.\n","\n","As you can see, several PDB files are indicated. The important columns to determine the best templates from this analysis are the fifth, sixth, seventh and eight columns, which correspond to the sequence length of the PDB hits and the target protein to be modelled, their sequence identity and e-value, respectively.\n","\n","**QUESTION:** From this analysis, which template would be better for modeling the structure of our target sequence?"]},{"cell_type":"markdown","metadata":{"id":"2ZxDoqLtt8Iz"},"source":["2. We will choose five PDB structures based on the sequence identity and e-value and select the most appropriate template for our target sequence among them. For this, we will first download these structures using the _Bio.PDB_ command from biopython, and then use the alignment.compare_structures() command to assess the structural and sequence similarity between the possible templates through the **compare.py** script.\n","\n","  Please take a few minutes to examine the content of this script, particularly i) how are the different protein structures included within the script; and ii) which chain is being used from each structure. This is important for cases where only one of many chains in the PDB corresponds to the protein. Also, please note that there are two alignment steps: first, a sequence alignment; second, a structural alignment."]},{"cell_type":"code","metadata":{"id":"OpQqQZk_wRcU"},"source":["#Downloading the PDB files using biopython\n","from Bio.PDB import *\n","templates = ['6eqe', '7dzv', '7ec8', '7nei', '6sbn']\n","pdbl = PDBList()\n","for s in templates:\n","  pdbl.retrieve_pdb_file(s, pdir='.', file_format =\"pdb\", overwrite=True)\n","  os.rename(\"pdb\"+s+\".ent\", s+\".pdb\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"XTZq7xCZuisR"},"source":["#Downloading the build_profile.py script from GitHub\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/scripts/compare.py\n","#Check this script and change the names of the PDB files if required"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"N5e2ij2dulhJ"},"source":["#Running the compare script\n","!mod10.4 compare.py\n","#Check the log file\n","!sed -ne '/Sequence identity comparison (ID_TABLE):/,$ p' compare.log"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"SaXEfXaMMpUY"},"source":["Several decisions can be made based on the results from this analysis, which were filtered in the previous code cell for simplicity.\n","\n","The resulting comparison between structures shows that PDB 6eqe has the highest resolution (0.9 √Ö). Second, 7ec8, 6sbn and 7nei form a separate group of structures from 6eqe and 6dzv. Third, with the exception of the comparison between 6eqe and 7dzv, the selected enzymes are quite different between each other in terms of their amino acid sequence, with the difference being on average 48% (sequence difference percentage corresponds to the numbers on the right-hand side of the graph).\n","\n","Thus, considering the sequence identity between the target sequence and 6eqe, its high resolution and other crystallographic quality features (that you can check on the **[PDB website](https://www.rcsb.org/)**), this is the best candidate for single-template modeling."]},{"cell_type":"markdown","metadata":{"id":"dt61bO8tOmrr"},"source":["3. Now, we will **align the sequence of our template protein with the sequence of our target protein**, such that we can model the structure.\n","\n","  How hard is it? Not at all! Just download the  **align2d.py** script into your working folder, check the script to verify how the sequence of the target and the protein structure are evoked, and execute the script as we have done before\n","\n"]},{"cell_type":"code","metadata":{"id":"wnpjKfnsTdln"},"source":["#Downloading the align2D.py script from GitHub\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/scripts/align2D.py\n","#Check this script and change the names of the PDB files if required"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"oC1m363k7WMB"},"source":["#Running the align2D script\n","!mod10.4 align2D.py"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"YkiMEeNSTbWm"},"source":["4. You will end up with two new files (aligned.ali and aligned.fasta) that contain the pairwise alignment of the target and template sequences. Load the FASTA file into [Alignment Viewer 2.0](https://fast.alignmentviewer.org/). You can also use our Colab-mounted MSA viewer below:"]},{"cell_type":"code","metadata":{"cellView":"form","id":"PEi1eCVZEffo"},"source":["#@title Protein MSA Viewer in Google Colab\n","#The following code is modified from the wonderful viewer developed by Damien Farrell\n","#https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner\n","\n","#Importing all modules first\n","import os, io, random\n","import string\n","import numpy as np\n","\n","from Bio.Seq import Seq\n","from Bio.Align import MultipleSeqAlignment\n","from Bio import AlignIO, SeqIO\n","\n","import panel as pn\n","import panel.widgets as pnw\n","pn.extension()\n","\n","from bokeh.plotting import figure\n","from bokeh.models import ColumnDataSource, Plot, Grid, Range1d\n","from bokeh.models.glyphs import Text, Rect\n","from bokeh.layouts import gridplot\n","\n","#Setting up the amino color code according to Zappo color scheme\n","def get_colors(seqs):\n","    #make colors for bases in sequence\n","    text = [i for s in list(seqs) for i in s]\n","    #Use Zappo color scheme\n","    clrs =  {'K':'red',\n","             'R':'red',\n","             'H':'red',\n","             'D':'green',\n","             'E':'green',\n","             'Q':'blue',\n","             'N':'blue',\n","             'S':'blue',\n","             'T':'blue',\n","             'A':'blue',\n","             'I':'blue',\n","             'L':'blue',\n","             'M':'blue',\n","             'V':'blue',\n","             'F':'orange',\n","             'Y':'orange',\n","             'W':'orange',\n","             'C':'blue',\n","             'P':'yellow',\n","             'G':'orange',\n","             '-':'white'}\n","    colors = [clrs[i] for i in text]\n","    return colors\n","\n","#Setting up the MSA viewer\n","def view_alignment(aln, fontsize=\"9pt\", plot_width=800):\n","    \"\"\"Bokeh sequence alignment view\"\"\"\n","\n","    #make sequence and id lists from the aln object\n","    seqs = [rec.seq for rec in (aln)]\n","    ids = [rec.id for rec in aln]\n","    text = [i for s in list(seqs) for i in s]\n","    colors = get_colors(seqs)\n","    N = len(seqs[0])\n","    S = len(seqs)\n","    width = .4\n","\n","    x = np.arange(1,N+1)\n","    y = np.arange(0,S,1)\n","    #creates a 2D grid of coords from the 1D arrays\n","    xx, yy = np.meshgrid(x, y)\n","    #flattens the arrays\n","    gx = xx.ravel()\n","    gy = yy.flatten()\n","    #use recty for rect coords with an offset\n","    recty = gy+.5\n","    h= 1/S\n","    #now we can create the ColumnDataSource with all the arrays\n","    source = ColumnDataSource(dict(x=gx, y=gy, recty=recty, text=text, colors=colors))\n","    plot_height = len(seqs)*15+50\n","    x_range = Range1d(0,N+1, bounds='auto')\n","    if N>100:\n","        viewlen=100\n","    else:\n","        viewlen=N\n","    #view_range is for the close up view\n","    view_range = (0,viewlen)\n","    tools=\"xpan, xwheel_zoom, reset, save\"\n","\n","    #entire sequence view (no text, with zoom)\n","    p = figure(title=None, width= plot_width, height=50,\n","               x_range=x_range, y_range=(0,S), tools=tools,\n","               min_border=0, toolbar_location='below')\n","    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n","                 line_color=None, fill_alpha=0.6)\n","    p.add_glyph(source, rects)\n","    p.yaxis.visible = False\n","    p.grid.visible = False\n","\n","    #sequence text view with ability to scroll along x axis\n","    p1 = figure(title=None, width=plot_width, height=plot_height,\n","                x_range=view_range, y_range=ids, tools=\"xpan,reset\",\n","                min_border=0, toolbar_location='below')#, lod_factor=1)\n","    glyph = Text(x=\"x\", y=\"y\", text=\"text\", text_align='center',text_color=\"black\",\n","                text_font=\"monospace\",text_font_size=fontsize)\n","    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n","                line_color=None, fill_alpha=0.4)\n","    p1.add_glyph(source, glyph)\n","    p1.add_glyph(source, rects)\n","\n","    p1.grid.visible = False\n","    p1.xaxis.major_label_text_font_style = \"bold\"\n","    p1.yaxis.minor_tick_line_width = 0\n","    p1.yaxis.major_tick_line_width = 0\n","\n","    p = gridplot([[p],[p1]], toolbar_location='below')\n","    return p\n","\n","#Loading the viewer by indicating the MSA file and format to read\n","#@markdown Name of the MSA file (including the filetype)\n","MSAfile = 'aligned.fasta' #@param {type:\"string\"}\n","MSAformat = 'fasta' #@param {type:\"string\"}\n","aln = AlignIO.read(MSAfile,MSAformat)\n","p = view_alignment(aln, plot_width=900)\n","pn.pane.Bokeh(p)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"z-mqexQVEcUW"},"source":["  You will see that something odd is happening: **a large segment of the N-terminal has no equivalent residues in the template structure!**"]},{"cell_type":"markdown","metadata":{"id":"ED6ukIVBU8K_"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/cm_03.png'/>\n","</center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"DJAYQEdgU65T"},"source":["  This exercise leads to a general recommendation: you should always check all possible information on your biological sequences and structures to identify if these conflicts have any biological explanation. **THIS IS FUNDAMENTAL**."]},{"cell_type":"markdown","metadata":{"id":"WH3xPCuaPIIq"},"source":["5. Go to the PDB website and check the annotation of the sequence of the structure that you are using as template (in this case, 6EQE) in **Protein Feature View**. What do you see?\n","\n","  Based on your observations, decide whether you should truncate the sequence of your target before modeling, edit your target sequence and repeat the alignment, if required.\n"]},{"cell_type":"markdown","metadata":{"id":"3so-qzfD-zCU"},"source":["#Part III - Generate and visualize a comparative model using MODELLER"]},{"cell_type":"markdown","metadata":{"id":"bOtoHH4rPRvr"},"source":["1. Once your target and template sequences are aligned, use the **model-single.py** script for finally obtaining a structure of your target through comparative modeling. Again, read the script and check how the sequences and structures are called in MODELLER through these scripts. In this case, we are also performing this step on a separate folder.\n","\n","  Please note that 1 model is not enough, as there is an energy function defining the optimal position of atomic coordinates, thus different models will have different energies. Generally, between 50-100 are generated for sufficient evaluation.\n","\n","**üí° HINT:** For our example, the generation of 50 models takes around 15 min on Google Colab, whereas 10 models are generated in about 3 min. You can edit the number of models to generate on the `model-single.py` script."]},{"cell_type":"code","metadata":{"id":"PueeTZCbBRS9"},"source":["#Creating a new folder and copying the required files for MODELLER\n","%mkdir model-single\n","%cd model-single\n","%cp ../6eqe.pdb .\n","%cp ../aligned.ali ."],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"34o7ei0ZTr89"},"source":["#Downloading the model-single.py script from GitHub\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/scripts/model-single.py\n","#Check this script and change the names of the PDB files if required"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"QycjAY0JAbio"},"source":["#Running the model-single script\n","!mod10.4 model-single.py"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"S-VeK98_PeES"},"source":["2. The output from this process is a bunch of PDB files, each one of them corresponding to a comparative model of our target protein, that are numbered from 1 up to the total number of models requested during comparative modeling.\n","\n","  Also, the **model-single.log** output has the total potential energy for each structure,according to MODELLER‚Äôs DOPE (discrete optimized protein energy) score. For simplicity, this script was modified to indicate the model with the best DOPE score. We will be working only with the model with the best score for the remainder of the session.\n","  \n","  As an example, our best model during preparation of this tutorial showed the following DOPE score:\n","\n","```\n","Top model: target.B99990025.pdb (DOPE score -28735.180)\n","```"]},{"cell_type":"markdown","metadata":{"id":"4z5y3ANOGczz"},"source":["3. Before we check the quality of our model, we will take a look at it on **py3Dmol**.\n","\n","**üí° HINT:** We are creating a copy of our model and changing the chain id from A to B, in order to load both structures into py3Dmol"]},{"cell_type":"code","metadata":{"id":"N_wN77GlGoV8"},"source":["#Copying our best model with a new chain id\n","!sed \"s/ A / B /g\" target.B99990004.pdb > bestmodel.pdb\n","\n","#Setting up py3Dmol for visualization\n","view=py3Dmol.view()\n","#Loading template\n","view.addModel(open('6eqe.pdb', 'r').read(),'pdb')\n","#Loading best DOPE score model\n","view.addModel(open('bestmodel.pdb', 'r').read(),'pdb')\n","#Coloring the structures by chain id\n","view.setStyle({'cartoon': {'colorscheme':'chain'}})\n","view.zoomTo()\n","view.setBackgroundColor('white')\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"LHbw7i6MGDNw"},"source":["4. Finally, to check the stereochemical quality of the model and its comparison to experimentally solved structures, we will use the [SAVES server](https://saves.mbi.ucla.edu), which employs several structure-based scoring strategies:\n","\n","* **VERIFY3D** (i.e. compatibility of an atomic 3D model to its 1D sequence when compared tothe energetics of good structures from the PDB).\n","* **ERRAT** (i.e. quality of non-bonded interactions of a region when compared to similar regions from highly refined structures).\n","* **PROCHECK** (stereochemical and geometrical quality of the model, via Ramachandran plots, sidechain rotamers, etc)."]},{"cell_type":"markdown","metadata":{"id":"BgjvtJccRs4-"},"source":["5. Download your best model, upload it to SAVES and wait for the results. Briefly:\n","- **Check the VERIFY3D results:** >80% of the residues should have an average score ‚â• 0.2, whereas the score profile allows you to identify conflicting regions.\n","- **Check the Ramachandran plot:** Are there any residues outside the allowed regions? What types of residues are found within those regions? (Check it by clicking on each dot in the plot)\n","- **Check the errors in PROCHECK:** are the errors located within the loop regions?\n"]},{"cell_type":"markdown","metadata":{"id":"yy_dPWhIVOCp"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/cm_04.png'/>\n","</center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"L57wIbIJYAiK"},"source":["While we will discuss some of these results at the end of this tutorial, we highly encourage you to read [this article](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007449), which contains more recommendations on comparative modelling of protein structures."]},{"cell_type":"markdown","metadata":{"id":"IBjMZe5ufofl"},"source":["**This is the end of the fourth tutorial! Good Science!!**"]}]}