{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"gDvK9xHtcjTr"},"source":["#Lab.05 / IBM3202 ‚Äì¬†Membrane Protein Modelling using PyRosetta"]},{"cell_type":"markdown","metadata":{"id":"CTFZC7Mafuzu"},"source":["##Theoretical Aspects"]},{"cell_type":"markdown","metadata":{"id":"VWTkCQifjWXY"},"source":["As stated by Alford et al. 2020 [Biophys J 2020 118(8), 2042-2055], membrane proteins partner with the surrounding lipid environment to perform essential life processes. They constitute 30% of all proteins and are targets for over 60% of pharmaceuticals. However, experimental difficulties have limited our insights into their molecular mechanisms of function."]},{"cell_type":"markdown","metadata":{"id":"0nPD5ByvctQy"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/mpmodelling_01.jpg'/>\n","<figcaption>FIGURE 1. The challenge of solving the structure of membrane proteins. As shown in this graph, the number of membrane proteins solved before the extensive use of cryoEM was (and still is) significantly inferior to the number of soluble proteins.\n","<br>Bijelic A & Rompel A (2015) <i>Coord Chem Rev 299, 22-38.\n","</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"iHpJP9wCdgqp"},"source":["Therefore, membrane protein modelling is ideal for determining the structure of these proteins and for developing new therapeutics. In this tutorial we will learn to model membrane proteins in PyRosetta and provide examples of the types of problems one can solve with this framework."]},{"cell_type":"markdown","metadata":{"id":"wr-vcvaVTKTf"},"source":["##Overview"]},{"cell_type":"markdown","metadata":{"id":"62vLxb1sSVVf"},"source":["In this tutorial, we will model different homologs with varying sequence identities of the membrane protein Glycophorin A, also known as GYPA. This is a pore-forming protein that, in the case of the human ortholog, has been broadly studied using different structural biology methods such as NMR and X-Ray crystallography.\n","\n","This overview and tutorial include information from this [tutorial](http://www.meilerlab.org/jobs/downloadfile/name/May2018_rosetta_cm_presentation.pdf) by Nica Marlow and Georg Kuenze"]},{"cell_type":"markdown","metadata":{"id":"jmsmt7SGbTmQ"},"source":["**The Algorithm**\n","\n","At a high-level, the algorithm consists of a long Monte Carlo trajectory starting from a randomly-chosen template. The MC trajectory employs the following moves:\n","\n","* Fragment insertion in unaligned regions\n","\n","* Replacement of a randomly-chosen segment with that from a different template structure\n","\n","* Cartesian-space minimization using a smooth (differentiable) version of the Rosetta centroid energy function. Finally, this is followed by all-atom optimization.\n","\n","See more at the [official docs](https://www.rosettacommons.org/docs/latest/application_documentation/structure_prediction/RosettaCM)"]},{"cell_type":"markdown","metadata":{"id":"_rm1_EjsTMWF"},"source":["**Differences between Single and Multiple Template modeling**\n","\n","Single Template Modeling:\n","- Single template as input\n","- Uses sequence and template derived\n","fragments\n","- Used when available templates have very\n","high identity (>60%)\n","\n","Multiple Template Modeling:\n","- Multiple templates as input\n","- Combine sections of multiple threaded\n","models and sequence derived fragments\n","- Used when available templates have low\n","identity (30-50%)"]},{"cell_type":"markdown","metadata":{"id":"eiINEFYXfajF"},"source":["**The protocol**\n","\n","Rosetta comparative modelling (**RosettaCM**) performs a complex series of steps that are illustrated in the scheme below, but that can be summarize in three main steps:\n","\n","1. Align target sequence to template sequences\n","\n","2. Perform a partial-thread of the target sequence onto template structures\n","\n","3. Combine pieces from different templates using RosettaCM hybridize"]},{"cell_type":"markdown","metadata":{"id":"NK5aEeL_e2w_"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/mpmodelling_02.jpg'/>\n","<figcaption>FIGURE 2. A simple scheme of the RosettaCM protocol\n","<br>Song Y et al (2013) <i>Structure 21(10), 1735-1742</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"eTZzTm6SUOIH"},"source":["**Input Files for RosettaCM**\n","\n","Bare minimum:\n","\n","* Partial-threaded structures (‚úÖ)\n","\n","* Mover definition and options (‚úÖ)\n","\n","Specific to membrane proteins (not needed if modeling\n","soluble proteins):\n","\n","* Membrane spanning regions (span file) (üôã)\n","\n","* Membrane weight patches  (‚úÖ)\n","\n","Optional files based on available information:\n","\n","* Constraint information (eg. atom pair connectivity)\n","\n","* Disulfide Connectivity"]},{"cell_type":"markdown","metadata":{"id":"SKd7eVaiRunj"},"source":["**For a deep dive into membrane protein modelling using Rosetta, we suggest to check the following articles:**\n","1. Alford RF, Leman JK et al., ‚ÄúAn Integrated Framework Advancing Membrane Protein Modeling and Design,‚Äù PLOS Computational Biology. 2015;11(9).\n","2. Alford RF, Fleming PJ, Fleming KG, Gray JJ (2020) ‚ÄúProtein structure prediction and design in a biologically-realistic implicit membrane‚Äù BioRxiv\n","3. Koehler Leman J, Muller BK, Gray JJ (2017) ‚ÄúExpanding the toolkit for membrane protein modeling in Rosetta‚Äù Bioinformatics 33(5):754-756.\n","4. Koehler Leman J & Gray JJ (2015) ‚ÄúComputational modeling of membrane proteins‚Äù Proteins Structure, Function, and Bioinformatics 83(1):1-24."]},{"cell_type":"markdown","metadata":{"id":"U2vWpWg8klhM"},"source":["#Part I ‚Äì Exploring the sequences of GYPA orthologs"]},{"cell_type":"markdown","metadata":{"id":"D-u6sgce8Zxm"},"source":["1. The sequences that we are going to be working with today  correspond to three ortologs of the Glycophorin A membrane protein. In our particular case we are going to model the following ortologs using PyRosetta:\n","\n","  *   [Marmota flaviventris](https://www.ncbi.nlm.nih.gov/protein/XP_027782564.1?report=fasta&log$=seqview)\n","\n","  *   [Gorilla gorilla](https://www.ncbi.nlm.nih.gov/protein/XP_007666620.2?report=fasta&log$=seqview)\n","  \n","  *   [Ornithorhynchus anatinus](https://www.uniprot.org/uniprot/O18798)"]},{"cell_type":"markdown","metadata":{"id":"xHLARFm0-8ae"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/mpmodelling_03.png'/>\n","</figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"ESDFFpOd-pYU"},"source":["2. For the sake of time we are going to provide the already aligned and trimmed sequences in a Grishin format file (yet another alignment format), the partial-threaded structures required to run the modelling protocol and the fragment files:\n","\n","  You will find these files inside its corresponding folder inside input_files. If you want to pursue the generation of these input files on your own, check **Appendix A**.\n","\n"]},{"cell_type":"code","metadata":{"id":"ImD_xOcegVuQ"},"source":["#Here we download the input files from Github\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/input_files/input_files_mp.zip\n","#This remove command is just in case you had some previous input files\n","!rm -rf input_files\n","#Here, we unzip the downloaded files\n","!unzip input_files_mp.zip"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"agqXm_0dmf8k"},"source":["#Part II ‚Äì Determining the membrane spanning regions (span file) using OCTOPUS server"]},{"cell_type":"markdown","metadata":{"id":"a0Qhw-MSmtjk"},"source":["As we briefly noted at the begining of this tutorial (Overview), we need to give Rosetta a way of knowing to which residues a specific and implicit \"membrane\" potential needs to be applied. These instructions are generated based on the transmembrane (TM) prediction server [**OCTOPUS**](https://octopus.cbr.su.se)"]},{"cell_type":"markdown","metadata":{"id":"ZIycpMeaCwI4"},"source":["1. In [**OCTOPUS**](https://octopus.cbr.su.se) you need to input the target protein sequence for the TM prediction. You can find these protein sequences inside the FASTA folder within the downloaded input files."]},{"cell_type":"markdown","metadata":{"id":"jNyDLK3CCShq"},"source":["2. The, the topology files must be processed to convert them into span files as shown in the code cell below:"]},{"cell_type":"code","metadata":{"id":"u_pxQELHbPX7"},"source":["%%bash\n","cd /content/input_files/octopus\n","perl octopus2span.pl GYPA.topo.txt > GYPA.span\n","perl octopus2span.pl marmota.topo.txt > marmota.span\n","perl octopus2span.pl ornithorhynchus.topo.txt > ornithorhynchus.span\n","perl octopus2span.pl gorilla.topo.txt > gorilla.span"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zjUSFLdJCY7f"},"source":["3. Once you obtained all the span files from **OCTOPUS**, remember to place them inside the following folder\n","\n","```\n","/content/input_files/octopus #Sweet trick! You can open the folder clicking this path\n","```\n","\n"]},{"cell_type":"markdown","metadata":{"id":"o3hcWpFQXoUx"},"source":["# Part III ‚Äì Setting Up PyRosetta and Executing the Membrane Protein Modelling protocols\n"]},{"cell_type":"markdown","metadata":{"id":"i4mcFtHR6u_-"},"source":["1. We start by setting up PyRosetta in Google Colab, which will be installed in your Google Drive, for which you must provide access."]},{"cell_type":"code","metadata":{"id":"SePtDf1JST8a"},"source":["!pip install pyrosettacolabsetup\n","import pyrosettacolabsetup; pyrosettacolabsetup.install_pyrosetta()\n","import pyrosetta; pyrosetta.init()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"MglIIrmAhWZ0"},"source":["2. In the following cell code, we import the PyRosetta modules that we are going to use"]},{"cell_type":"code","metadata":{"id":"e02ruAYR076e"},"source":["#Here we import the PyRosetta modules that we are going to be using\n","%%time\n","import pandas as pd\n","from pyrosetta import *\n","from pyrosetta.rosetta.protocols.rosetta_scripts import *\n","from pyrosetta import (\n","    init, pose_from_sequence, pose_from_file, Pose, MoveMap, create_score_function, get_fa_scorefxn,\n","    MonteCarlo, TrialMover, SwitchResidueTypeSetMover, PyJobDistributor,\n",")\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4vFqicSeD-NQ"},"source":["3. Next, we performing the Membrane Protein Modelling protocol. This is briefly described for the GYPA from _Gorilla_, whereas the modelling of GYPA for the other orthologs is just a modification of this protocol."]},{"cell_type":"markdown","metadata":{"id":"wlmDdcxmXZ9T"},"source":["#### Gorilla ü¶ç"]},{"cell_type":"markdown","metadata":{"id":"cQc4_jK8Oug2"},"source":["1. In the following cell code we give PyRosetta all the needed **Flags**. These flags are equivalent to command-line instructions in order to set the path for files and set the value of different (and numerous) options. We pass them into PyRosetta, thus initializing it\n","\n","\n","```\n","pyrosetta.init(flags)\n","```\n","\n"]},{"cell_type":"code","metadata":{"id":"vG9dJLT4bZkk"},"source":["#Flags are equivalent to command-line instructions in order to set the path for files and set the value of different options\n","flags = f\"\"\"\n","-no_fconfig\n","-in:file:fasta ./input_files/fasta/gorillaGYPA.fasta\n","-out:prefix GYPA_gorilla_\n","-out:path:all ./output_files\n","-in:file:spanfile ./input_files/octopus/gorilla.span\n","-membrane:no_interpolate_Mpair\n","-membrane:Menv_penalties\n","-rg_reweight .1\n","-relax:fast\n","-relax:minimize_bond_angles\n","-relax:minimize_bond_lengths\n","-relax:jump_move true\n","-default_max_cycles 2\n","-relax:min_type lbfgs_armijo_nonmonotone\n","-score:weights input_files/wts/stage3_rlx_membrane.wts\n","-use_bicubic_interpolation\n","-hybridize:stage1_probability 1.0\n","-sog_upper_bound 15\n","-chemical:exclude_patches LowerDNA  UpperDNA Cterm_amidation SpecialRotamer VirtualBB ShoveBB VirtualDNAPhosphate VirtualNTerm CTermConnect sc_orbitals pro_hydroxylated_case1 pro_hydroxylated_case2 ser_phosphorylated thr_phosphorylated  tyr_phosphorylated tyr_sulfated lys_dimethylated lys_monomethylated  lys_trimethylated lys_acetylated glu_carboxylated cys_acetylated tyr_diiodinated N_acetylated C_methylamidated MethylatedProteinCterm\n","-analytic_etable_evaluation true\n","-linmem_ig 10\n","-out:mute core.io.pose_from_sfr.PoseFromSFRBuilder core.scoring.MembranePotential core.chemical.GlobalResidueTypeSet basic.io.database core.scoring.MembraneTopology core.scoring.CartesianBondedEnergy core.conformation.Conformation protocols.jd2.parser.ScoreFunctionLoader core.optimization.AtomTreeMinimizer core.pack\n","\"\"\"\n","pyrosetta.init(flags)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"3EwFD1YdO-xV"},"source":["2. One of the crucial concepts in PyRosetta are **poses**. They contain the information, sequence or structure, to which the protocol will be applied.\n","\n","  With this in mind, here we create a **pose object** from a sequence, which in this case is the sequence of the GYPA from **_Gorilla_**. Also, we specify a **res_type** which corresponds to the **fullatom_standard**."]},{"cell_type":"code","metadata":{"id":"MQRxSQh4bZku"},"source":["#Here we create a pose object from a sequence\n","#Also, we specify a res_type which corresponds to the fullatom_standard\n","%%time\n","pose = pose_from_sequence('DNGEWIQLVPRFSGPEITLIIFGVMAGIIGTILLISYSIRRL', 'fa_standard')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"40X9c4WjPXeq"},"source":["3. While there are multiple ways of creating a protocol in Rosetta, in our case we are going to create a protocol from a **XML object**.\n","\n","**NOTE:** In the **MOVERS** tag we specify the **fragment files** and the **threaded_pdb**!!"]},{"cell_type":"code","metadata":{"id":"pLfWzrs0bZky"},"source":["#Creating a Rosetta protocol from a XML object\n","#Note that in the <MOVERS> we specify the fragment files and the threaded_pdb!!\n","\n","%%time\n","xml = pyrosetta.rosetta.protocols.rosetta_scripts.XmlObjects.create_from_string(\"\"\"\n","<ROSETTASCRIPTS>\n","    <TASKOPERATIONS>\n","    </TASKOPERATIONS>\n","    <RESIDUE_SELECTORS>\n","    </RESIDUE_SELECTORS>\n","    <SCOREFXNS>\n","      <ScoreFunction name=\"stage1\" weights=\"input_files/wts/stage1_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"1\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"stage2\" weights=\"input_files/wts/stage2_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"fullatom\" weights=\"input_files/wts/stage3_rlx_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"membrane\" weights=\"membrane_highres_Menv_smooth\" symmetric=\"0\">\n","        <Reweight scoretype=\"cart_bonded\" weight=\"0.5\"/>\n","        <Reweight scoretype=\"pro_close\" weight=\"0\"/>\n","\t <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","\t</ScoreFunction>\n","    </SCOREFXNS>\n","    <FILTERS>\n","    </FILTERS>\n","    <MOVERS>\n","        \t <Hybridize name=\"hybridize\" stage1_scorefxn=\"stage1\" stage2_scorefxn=\"stage2\" fa_scorefxn=\"fullatom\" batch=\"1\" stage1_increase_cycles=\"0.5\" stage2_increase_cycles=\"0.5\" linmin_only=\"1\" add_hetatm=\"0\" hetatm_to_protein_cst_weight=\"0\"  >\n","       \t\t\t<Fragments three_mers=\"/content/input_files/frags/gorillaGYPA.9mers\" nine_mers=\"/content/input_files/frags/gorillaGYPA.9mers\" />\n","\t          \t<Template pdb=\"/content/input_files/threaded_pdbs/5eh6_A.gorilla.pdb\" cst_file=\"AUTO\" weight=\"1.000\" />\n","\t       </Hybridize>\n","    </MOVERS>\n","    <APPLY_TO_POSE>\n","    </APPLY_TO_POSE>\n","    <PROTOCOLS>\n","      Add mover=\"add_memb\"\n","\t    <Add mover=\"hybridize\"/>\n","    </PROTOCOLS>\n","<OUTPUT scorefxn=\"membrane\"/>\n","</ROSETTASCRIPTS>\n","    \"\"\").get_mover(\"ParsedProtocol\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Dn9Gu8TkjmVU"},"source":["4. Once our protocol is set up, in the following code cells:\n","  * We create a membrane score function\n","  * Then, we call the PyJobDistrubutor and pass three arguments\n","    * A name prefix to output our models as PDB files\n","    * The number of models to generate\n","    * The score function to employ\n","  * We pass our pose to the job distributor that we just created\n","  * Then, we create a Pandas dataframe to store the score results\n","  * Finally, we loop while the jobs are not completed"]},{"cell_type":"code","metadata":{"id":"flBA8JKLbZk6"},"source":["#Here we create a membrane score function\n","membrane = pyrosetta.create_score_function(\"membrane_highres_Menv_smooth\")\n","#Here we create a job distributor object and specify the prefix, number of output models or poses and the scorefunction\n","jd = pyrosetta.toolbox.py_jobdistributor.PyJobDistributor(pdb_name=\"GYPA_gorilla_\",nstruct=2,scorefxn=membrane)\n","#Here we pass our previously created pose to the Job Distributor\n","jd.native_pose = pose\n","#Here we create a panda dataframe in order to store our score results as we are going to see later\n","gdf = pd.DataFrame()\n","#Here we loop until all n jobs are completed\n","while not jd.job_complete:\n","    test_pose = pose.clone()\n","    xml.apply(test_pose)\n","    test_gdf = pd.DataFrame.from_records(dict(test_pose.scores), index=[jd.current_name])\n","    gdf = pd.concat([gdf, test_gdf])\n","    jd.output_decoy(test_pose)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Mh_oZyC4XgY6"},"source":["#### Ornitorrinchus ü•ö"]},{"cell_type":"code","metadata":{"id":"PmDCmCogbbja"},"source":["flags = f\"\"\"\n","-no_fconfig\n","-in:file:fasta ./input_files/fasta/ornithorhynchusGYPA.fasta\n","-out:prefix GYPA_ornithorhynchus_\n","-in:file:spanfile ./input_files/octopus/ornithorhynchus.span\n","-membrane:no_interpolate_Mpair\n","-membrane:Menv_penalties\n","-rg_reweight .1\n","-relax:fast\n","-relax:minimize_bond_angles\n","-relax:minimize_bond_lengths\n","-relax:jump_move true\n","-default_max_cycles 2\n","-relax:min_type lbfgs_armijo_nonmonotone\n","-score:weights input_files/wts/stage3_rlx_membrane.wts\n","-use_bicubic_interpolation\n","-hybridize:stage1_probability 1.0\n","-sog_upper_bound 15\n","-chemical:exclude_patches LowerDNA  UpperDNA Cterm_amidation SpecialRotamer VirtualBB ShoveBB VirtualDNAPhosphate VirtualNTerm CTermConnect sc_orbitals pro_hydroxylated_case1 pro_hydroxylated_case2 ser_phosphorylated thr_phosphorylated  tyr_phosphorylated tyr_sulfated lys_dimethylated lys_monomethylated  lys_trimethylated lys_acetylated glu_carboxylated cys_acetylated tyr_diiodinated N_acetylated C_methylamidated MethylatedProteinCterm\n","-analytic_etable_evaluation true\n","-linmem_ig 10\n","-out:mute core.io.pose_from_sfr.PoseFromSFRBuilder core.scoring.MembranePotential core.chemical.GlobalResidueTypeSet basic.io.database core.scoring.MembraneTopology core.scoring.CartesianBondedEnergy core.conformation.Conformation protocols.jd2.parser.ScoreFunctionLoader core.optimization.AtomTreeMinimizer core.pack\n","\"\"\"\n","pyrosetta.init(flags)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"NxxxpSsKbbjd"},"source":["%%time\n","pose = pose_from_sequence('IDPEIEHIFTGPVAAVIIYAVVCGVIGTILFIALVIKVV', 'fa_standard')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"XKkrl3q8bbjf"},"source":["%%time\n","xml = pyrosetta.rosetta.protocols.rosetta_scripts.XmlObjects.create_from_string(\"\"\"\n","<ROSETTASCRIPTS>\n","    <TASKOPERATIONS>\n","    </TASKOPERATIONS>\n","    <RESIDUE_SELECTORS>\n","    </RESIDUE_SELECTORS>\n","    <SCOREFXNS>\n","      <ScoreFunction name=\"stage1\" weights=\"input_files/wts/stage1_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"1\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"stage2\" weights=\"input_files/wts/stage2_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"fullatom\" weights=\"input_files/wts/stage3_rlx_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"membrane\" weights=\"membrane_highres_Menv_smooth\" symmetric=\"0\">\n","        <Reweight scoretype=\"cart_bonded\" weight=\"0.5\"/>\n","        <Reweight scoretype=\"pro_close\" weight=\"0\"/>\n","\t <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","\t</ScoreFunction>\n","    </SCOREFXNS>\n","    <FILTERS>\n","    </FILTERS>\n","    <MOVERS>\n","          AddMembraneMover name=\"add_memb\"\n","        \t <Hybridize name=\"hybridize\" stage1_scorefxn=\"stage1\" stage2_scorefxn=\"stage2\" fa_scorefxn=\"fullatom\" batch=\"1\" stage1_increase_cycles=\"0.5\" stage2_increase_cycles=\"0.5\" linmin_only=\"1\" add_hetatm=\"0\" hetatm_to_protein_cst_weight=\"0\"  >\n","       \t\t\t<Fragments three_mers=\"/content/input_files/frags/ornithorhynchusGYPA.3mers\" nine_mers=\"/content/input_files/frags/ornithorhynchusGYPA.9mers\" />\n","\t          \t<Template pdb=\"/content/input_files/threaded_pdbs/5eh6_A.ornithorhynchus.pdb\" cst_file=\"AUTO\" weight=\"1.000\" />\n","\t       </Hybridize>\n","    </MOVERS>\n","    <APPLY_TO_POSE>\n","    </APPLY_TO_POSE>\n","    <PROTOCOLS>\n","      Add mover=\"add_memb\"\n","\t    <Add mover=\"hybridize\"/>\n","    </PROTOCOLS>\n","<OUTPUT scorefxn=\"membrane\"/>\n","</ROSETTASCRIPTS>\n","    \"\"\").get_mover(\"ParsedProtocol\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_a7SPqrMbbji"},"source":["%%time\n","membrane = pyrosetta.create_score_function(\"membrane_highres_Menv_smooth\")\n","jd = pyrosetta.toolbox.py_jobdistributor.PyJobDistributor(pdb_name=\"GYPA_ornitorrinchus\",nstruct=2,scorefxn=membrane)\n","jd.native_pose = pose\n","odf = pd.DataFrame()\n","while not jd.job_complete:\n","    test_pose = pose.clone()\n","    xml.apply(test_pose)\n","    test_odf = pd.DataFrame.from_records(dict(test_pose.scores), index=[jd.current_name])\n","    odf = pd.concat([odf, test_odf])\n","    jd.output_decoy(test_pose)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"feibleVGXjZX"},"source":["####Marmota ü¶°"]},{"cell_type":"code","metadata":{"id":"sh2p_Qq0bee4"},"source":["%%time\n","flags = f\"\"\"\n","-no_fconfig\n","-in:file:fasta ./input_files/fasta/marmotaGYPA.fasta\n","-out:prefix GYPA_marmota_\n","-out:path:all ./output_files\n","-in:file:spanfile ./input_files/octopus/marmota.span\n","-membrane:no_interpolate_Mpair\n","-membrane:Menv_penalties\n","-rg_reweight .1\n","-relax:fast\n","-relax:minimize_bond_angles\n","-relax:minimize_bond_lengths\n","-relax:jump_move true\n","-default_max_cycles 2\n","-relax:min_type lbfgs_armijo_nonmonotone\n","-score:weights input_files/wts/stage3_rlx_membrane.wts\n","-use_bicubic_interpolation\n","-hybridize:stage1_probability 1.0\n","-sog_upper_bound 15\n","-chemical:exclude_patches LowerDNA  UpperDNA Cterm_amidation SpecialRotamer VirtualBB ShoveBB VirtualDNAPhosphate VirtualNTerm CTermConnect sc_orbitals pro_hydroxylated_case1 pro_hydroxylated_case2 ser_phosphorylated thr_phosphorylated  tyr_phosphorylated tyr_sulfated lys_dimethylated lys_monomethylated  lys_trimethylated lys_acetylated glu_carboxylated cys_acetylated tyr_diiodinated N_acetylated C_methylamidated MethylatedProteinCterm\n","-analytic_etable_evaluation true\n","-linmem_ig 10\n","-out:mute core.io.pose_from_sfr.PoseFromSFRBuilder core.scoring.MembranePotential core.chemical.GlobalResidueTypeSet basic.io.database core.scoring.MembraneTopology core.scoring.CartesianBondedEnergy core.conformation.Conformation protocols.jd2.parser.ScoreFunctionLoader core.optimization.AtomTreeMinimizer core.pack\n","\"\"\"\n","pyrosetta.init(flags)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GahGjthKbee7"},"source":["%%time\n","pose = pose_from_sequence('KQGRNQIIHPFSEPVIILIIFAVMFGIIGTILLISFCVRRL', 'fa_standard')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"BhLwUqtBbee9"},"source":["%%time\n","xml = pyrosetta.rosetta.protocols.rosetta_scripts.XmlObjects.create_from_string(\"\"\"\n","<ROSETTASCRIPTS>\n","    <TASKOPERATIONS>\n","    </TASKOPERATIONS>\n","    <RESIDUE_SELECTORS>\n","    </RESIDUE_SELECTORS>\n","    <SCOREFXNS>\n","      <ScoreFunction name=\"stage1\" weights=\"input_files/wts/stage1_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"1\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"stage2\" weights=\"input_files/wts/stage2_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"fullatom\" weights=\"input_files/wts/stage3_rlx_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"membrane\" weights=\"membrane_highres_Menv_smooth\" symmetric=\"0\">\n","        <Reweight scoretype=\"cart_bonded\" weight=\"0.5\"/>\n","        <Reweight scoretype=\"pro_close\" weight=\"0\"/>\n","\t <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","\t</ScoreFunction>\n","    </SCOREFXNS>\n","    <FILTERS>\n","    </FILTERS>\n","    <MOVERS>\n","          AddMembraneMover name=\"add_memb\"\n","        \t <Hybridize name=\"hybridize\" stage1_scorefxn=\"stage1\" stage2_scorefxn=\"stage2\" fa_scorefxn=\"fullatom\" batch=\"1\" stage1_increase_cycles=\"0.5\" stage2_increase_cycles=\"0.5\" linmin_only=\"1\" add_hetatm=\"0\" hetatm_to_protein_cst_weight=\"0\"  >\n","       \t\t\t<Fragments three_mers=\"./input_files/frags/marmotaGYPA.3mers\" nine_mers=\"./input_files/frags/marmotaGYPA.9mers\" />\n","\t          \t<Template pdb=\"./input_files/threaded_pdbs/5eh6_A.marmota.pdb\" cst_file=\"AUTO\" weight=\"1.000\" />\n","\t       </Hybridize>\n","    </MOVERS>\n","    <APPLY_TO_POSE>\n","    </APPLY_TO_POSE>\n","    <PROTOCOLS>\n","      Add mover=\"add_memb\"\n","\t    <Add mover=\"hybridize\"/>\n","    </PROTOCOLS>\n","<OUTPUT scorefxn=\"membrane\"/>\n","</ROSETTASCRIPTS>\n","    \"\"\").get_mover(\"ParsedProtocol\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"l2l4g1onbefA"},"source":["membrane = pyrosetta.create_score_function(\"membrane_highres_Menv_smooth\")\n","jd = pyrosetta.toolbox.py_jobdistributor.PyJobDistributor(pdb_name=\"GYPA_marmota_\",nstruct=2,scorefxn=membrane)\n","jd.native_pose = pose\n","mdf = pd.DataFrame()\n","while not jd.job_complete:\n","    test_pose = pose.clone()\n","    xml.apply(test_pose)\n","    test_mdf = pd.DataFrame.from_records(dict(test_pose.scores), index=[jd.current_name])\n","    mdf = pd.concat([mdf, test_mdf])\n","    jd.output_decoy(test_pose)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"0ehjClu0YhXk"},"source":["####Human üßë"]},{"cell_type":"code","metadata":{"id":"slUPRFWuoOQR"},"source":["%%time\n","flags = f\"\"\"\n","-no_fconfig\n","-in:file:fasta ./input_files/fasta/GYPA.fasta\n","-parser:protocol ./input_files/cm.xml\n","-out:prefix GYPA_\n","-out:path:all ./output_files\n","-in:file:spanfile ./input_files/octopus/GYPA.span\n","-membrane:no_interpolate_Mpair\n","-membrane:Menv_penalties\n","-rg_reweight .1\n","-relax:fast\n","-relax:minimize_bond_angles\n","-relax:minimize_bond_lengths\n","-relax:jump_move true\n","-default_max_cycles 2\n","-relax:min_type lbfgs_armijo_nonmonotone\n","-score:weights input_files/wts/stage3_rlx_membrane.wts\n","-use_bicubic_interpolation\n","-hybridize:stage1_probability 1.0\n","-sog_upper_bound 15\n","-chemical:exclude_patches LowerDNA  UpperDNA Cterm_amidation SpecialRotamer VirtualBB ShoveBB VirtualDNAPhosphate VirtualNTerm CTermConnect sc_orbitals pro_hydroxylated_case1 pro_hydroxylated_case2 ser_phosphorylated thr_phosphorylated  tyr_phosphorylated tyr_sulfated lys_dimethylated lys_monomethylated  lys_trimethylated lys_acetylated glu_carboxylated cys_acetylated tyr_diiodinated N_acetylated C_methylamidated MethylatedProteinCterm\n","-analytic_etable_evaluation true\n","-linmem_ig 10\n","-out:mute core.io.pose_from_sfr.PoseFromSFRBuilder core.scoring.MembranePotential core.chemical.GlobalResidueTypeSet basic.io.database core.scoring.MembraneTopology core.scoring.CartesianBondedEnergy core.conformation.Conformation protocols.jd2.parser.ScoreFunctionLoader core.optimization.AtomTreeMinimizer core.pack\n","\"\"\"\n","pyrosetta.init(flags)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"dYwsIljin-vA"},"source":["%%time\n","pose = pose_from_sequence('ETGERVQLAHHFSEPEITLIIFGVMAGVIGTILLISYGIRRL', 'fa_standard')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"oYRxZf4M8_Un"},"source":["%%time\n","xml = pyrosetta.rosetta.protocols.rosetta_scripts.XmlObjects.create_from_string(\"\"\"\n","<ROSETTASCRIPTS>\n","    <TASKOPERATIONS>\n","    </TASKOPERATIONS>\n","    <RESIDUE_SELECTORS>\n","    </RESIDUE_SELECTORS>\n","    <SCOREFXNS>\n","      <ScoreFunction name=\"stage1\" weights=\"input_files/wts/stage1_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"1\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"stage2\" weights=\"input_files/wts/stage2_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"fullatom\" weights=\"input_files/wts/stage3_rlx_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"membrane\" weights=\"membrane_highres_Menv_smooth\" symmetric=\"0\">\n","        <Reweight scoretype=\"cart_bonded\" weight=\"0.5\"/>\n","        <Reweight scoretype=\"pro_close\" weight=\"0\"/>\n","\t <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","\t</ScoreFunction>\n","    </SCOREFXNS>\n","    <FILTERS>\n","    </FILTERS>\n","    <MOVERS>\n","          AddMembraneMover name=\"add_memb\"\n","        \t <Hybridize name=\"hybridize\" stage1_scorefxn=\"stage1\" stage2_scorefxn=\"stage2\" fa_scorefxn=\"fullatom\" batch=\"1\" stage1_increase_cycles=\"0.5\" stage2_increase_cycles=\"0.5\" linmin_only=\"1\" add_hetatm=\"0\" hetatm_to_protein_cst_weight=\"0\"  >\n","       \t\t\t<Fragments three_mers=\"./input_files/frags/GYPA.3mers\" nine_mers=\"./input_files/frags/GYPA.9mers\" />\n","\t          \t<Template pdb=\"/content/input_files/threaded_pdbs/5eh6_A.human.pdb\" cst_file=\"AUTO\" weight=\"1.000\" />\n","\t       </Hybridize>\n","    </MOVERS>\n","    <APPLY_TO_POSE>\n","    </APPLY_TO_POSE>\n","    <PROTOCOLS>\n","      Add mover=\"add_memb\"\n","\t    <Add mover=\"hybridize\"/>\n","    </PROTOCOLS>\n","<OUTPUT scorefxn=\"membrane\"/>\n","</ROSETTASCRIPTS>\n","    \"\"\").get_mover(\"ParsedProtocol\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"gvyV2upvoAW1"},"source":["%%time\n","membrane = pyrosetta.create_score_function(\"membrane_highres_Menv_smooth\")\n","jd = pyrosetta.toolbox.py_jobdistributor.PyJobDistributor(pdb_name=\"GYPA_human_\",nstruct=2,scorefxn=membrane)\n","jd.native_pose = pose\n","#Name of the score dataframe\n","hdf = pd.DataFrame()\n","while not jd.job_complete:\n","    test_pose = pose.clone()\n","    xml.apply(test_pose)\n","    test_hdf = pd.DataFrame.from_records(dict(test_pose.scores), index=[jd.current_name])\n","    hdf = pd.concat([hdf, test_hdf])\n","    jd.output_decoy(test_pose)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"1GSWo_C0qE3M"},"source":["#Part IV ‚Äì Analyze and visualize the modelled structures"]},{"cell_type":"markdown","metadata":{"id":"eVDe0mz3QRQi"},"source":["1. Now, we need to determine our best models by analyzing their score dataframes. For this, we are going to obtain the data frame for each of the models that we generated with PyRosetta.\n","\n","  We can visually inspect these data frames just writing their name in a code chunk."]},{"cell_type":"code","metadata":{"id":"sD8nrCKprPpr"},"source":["#Call here the name of the different dataframes that were created\n","hdf\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GZ1NnFdOlIT0"},"source":["odf\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"AixtSW1JlJXz"},"source":["mdf\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"yvfyIrqClKJL"},"source":["gdf"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gKSGvDM8Okly"},"source":["2. As Rosetta uses an stochastic algorithm for comparative modelling, we are only interested in the low energy structures. Thus, we can sort our models by the column total_score as in the following cell code:"]},{"cell_type":"code","metadata":{"id":"0XFEzgf4Li4f"},"source":["#hdf.sort_values('total_score')\n","gdf.sort_values('total_score')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"jE8R-lrqRdXe"},"source":["3. Having too many tables or tables with too much information would be challenging to analyze our results. For simplicity, we can make a plot with the **total_score** of each PDB for each species using **matplotlib**"]},{"cell_type":"code","metadata":{"id":"WU7m7LhCq8DG"},"source":["import matplotlib\n","import seaborn\n","frames = [hdf,odf,mdf,gdf]\n","results = pd.concat(frames)\n","matplotlib.rcParams['figure.figsize'] = [12.0, 8.0]\n","seaborn.barplot(x=\"total_score\", y=\"index\", data=results.sort_values('total_score').reset_index())"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Tc7oifPnbssg"},"source":["4. Finally, we can take a look at our best models using **py3Dmol**"]},{"cell_type":"code","metadata":{"id":"LPpEhymrJ5gM"},"source":["#!pip install py3Dmol\n","import py3Dmol\n","view=py3Dmol.view()\n","view.setBackgroundColor('white')\n","#Loading template\n","view.addModel(open('input_files/pdbs/5eh6.pdb', 'r').read(),'pdb')\n","view.addModel(open('GYPA_gorilla__0.pdb', 'r').read(),'pdb')\n","view.setStyle({'chain':'A'},{'cartoon': {'color':'spectrum'}})\n","view.zoomTo()\n","view.show()\n","#The following is a small trick to be able to manipulate (rotate, zoom) your\n","#representation in Google Colab and, once you are happy with the visualization,\n","#press Enter to print out as an onscreen png image file that you can download\n","#input()\n","#view.png()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"JFpJRYPwJ8w6"},"source":["**üìöHOMEWORK:** As a challenge, using what you have learned in the previous tutorials to superimpose structures and determine their similarity using RMSD, take your best structures and compare them to answer the following question: **Is the TM region different between orthologs?**"]},{"cell_type":"markdown","metadata":{"id":"7CV7AsCQIFrB"},"source":["5. To finalize, you can download your output files as indicated in the code cell below"]},{"cell_type":"code","metadata":{"id":"oy4qyfAGMDFp"},"source":["#download your output\n","%%bash\n","zip -r example.zip ./outs\n","mkdir outs\n","cp *.pdb ./outs/"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"lJpeTpYPIKfB"},"source":["**This is the end of the fifth tutorial. Good Science!**"]},{"cell_type":"markdown","metadata":{"id":"s9giUnnEAast"},"source":["#Appendix A"]},{"cell_type":"markdown","metadata":{"id":"n1NoW-mIX3Fp"},"source":["## Generating the input files using the standalone command-line version of Rosetta"]},{"cell_type":"markdown","metadata":{"id":"FF5pgAKmh14Y"},"source":["First, you need to get a Rosetta license [HERE](https://els2.comotion.uw.edu/product/rosetta), as you did for PyRosetta"]},{"cell_type":"markdown","metadata":{"id":"pZ8OMgudYJi9"},"source":["### Download and extract Rosetta in Google Colab (ETA ~15min)"]},{"cell_type":"code","metadata":{"id":"9eLMeKl_vyuf"},"source":["!wget --user=Academic_User --password=Xry3x4 https://www.rosettacommons.org/downloads/academic/3.12/rosetta_bin_linux_3.12_bundle.tgz\n","!tar xvf /content/rosetta_bin_linux_3.12_bundle.tgz"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"5XUiyaTK6nxy"},"source":["!unzip input_files.zip"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"oovXJfZUl5g7"},"source":["### Grishin File Generation"]},{"cell_type":"markdown","metadata":{"id":"ZaQNZszVl8bf"},"source":["The Grishin alignment file format is documented [here](https://www.rosettacommons.org/docs/latest/rosetta_basics/file_types/Grishan-format-alignment). The following is an example of its structure\n","```\n","## target_name template_name.pdb\n","#\n","scores from program: 0\n","0 HERE_GOES_YOUR_TARGET_SEQUENCE_ALIGNMENT_IN_ONE_LINE\n","0 HERE_GOES_YOUR_TEMPLATE_SEQUENCE_ALIGNMENT_IN_ONE_LINE\n","```\n","This is one of the real Grishin files used in this tutorial\n","```\n","## gorillaGYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 DNGEWIQLVPRFSGPEITLIIFGVMAGIIGTILLISYSIRRL\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","```\n","\n"]},{"cell_type":"code","metadata":{"id":"RmZi9ejLmvYF"},"source":["#Modify and run this cell to create grishin from clustal alignment\n","%%bash\n","echo \\\n","\"\"\"\\\n","## gorillaGYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 DNGEWIQLVPRFSGPEITLIIFGVMAGIIGTILLISYSIRRL\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","\"\"\" > /content/input_files/grishin/5eh6_on_gorillaGYPA.grishin\n","\n","echo \\\n","\"\"\"\\\n","## GYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 ETGERVQLAHHFSEPEITLIIFGVMAGVIGTILLISYGIRRL\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","\"\"\" > /content/input_files/grishin/5eh6_on_humanGYPA.grishin\n","\n","echo \\\n","\"\"\"\\\n","## marmotaGYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 -KQGRNQIIHPFSEPVIILIIFAVMFGIIGTILLISFCVRRL\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","\"\"\" > /content/input_files/grishin/5eh6_on_marmotaGYPA.grishin\n","\n","\n","echo \\\n","\"\"\"\\\n","## OrnithorhynchusGYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 ---IDPEIEHIFTGPVAAVIIYAVVCGVIGTILFIALVIKVV\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","\"\"\" > /content/input_files/grishin/5eh6_on_ornithorhynchusGYPA.grishin"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"iAfdLyUYYV40"},"source":["### Partial Threading"]},{"cell_type":"code","metadata":{"id":"COXQteMfi0Ku"},"source":["import urllib.request\n","urllib.request.urlretrieve('http://files.rcsb.org/download/5eh6.pdb', '/content/input_files/pdbs/5eh6.pdb')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"z6Z-k7TDigtJ"},"source":["%%bash\n","cd /content/input_files/pdbs\n","/content/rosetta_bin_linux_2020.08.61146_bundle/tools/protein_tools/scripts/clean_pdb.py 5eh6.pdb A\n","mv 5eh6_A.* ../cleaned"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"Yw46R_v1560X"},"source":["!/content/rosetta_bin_linux_2020.08.61146_bundle/main/source/bin/partial_thread.static.linuxgccrelease -in:file:fasta ./fasta/GYPA.fasta -in:file:alignment ./grishin/5eh6.grishin -in:file:template_pdb ./cleaned/5eh6_A.pdb\n","!mv 5eh6_A.pdb.pdb 5eh6_A.human.pdb\n","!/content/rosetta_bin_linux_2020.08.61146_bundle/main/source/bin/partial_thread.static.linuxgccrelease -in:file:fasta ./fasta/ornithorhynchusGYPA.fasta -in:file:alignment ./grishin/5eh6_on_ornithorhynchusGYPA.grishin -in:file:template_pdb ./cleaned/5eh6_A.pdb\n","!mv 5eh6_A.pdb.pdb 5eh6_A.ornithorhynchus.pdb\n","!/content/rosetta_bin_linux_2020.08.61146_bundle/main/source/bin/partial_thread.static.linuxgccrelease -in:file:fasta ./fasta/marmotaGYPA.fasta -in:file:alignment ./grishin/5eh6_on_marmotaGYPA.grishin -in:file:template_pdb ./cleaned/5eh6_A.pdb\n","!mv 5eh6_A.pdb.pdb 5eh6_A.gorilla.pdb\n","!/content/rosetta_bin_linux_2020.08.61146_bundle/main/source/bin/partial_thread.static.linuxgccrelease -in:file:fasta ./fasta/gorillaGYPA.fasta -in:file:alignment ./grishin/5eh6_on_gorillaGYPA.grishin -in:file:template_pdb ./cleaned/5eh6_A.pdb\n","!mv 5eh6_A.pdb.pdb 5eh6_A.marmota.pdb\n","!mv 5eh6_A.*pdb ./input_files/threaded_pdbs/"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2Wn-5LuU44Wm"},"source":["### Getting the fragments from oldRobetta server\n"]},{"cell_type":"markdown","metadata":{"id":"LmPktANT5uEe"},"source":["As stated in the Rosetta documentation fragments are:\n","\n","\n","> (...) used in the assembly of proteins whether for structure prediction or design, to cut down on the size of the protein-folding search space. They are a core part of the Rosetta design.\n","\n","For the purpose of this tutorial, we skipped this part. However, in the following figure you can see a screenshot of the [Robetta-server](http://old.robetta.org/fragmentsubmit.jsp) where you can obtain the fragments. For this, you have to register and submit a FASTA sequence of your target proteins. Do keep in mind that the length of your target will impact in the time needed for the fragment picking process."]},{"cell_type":"markdown","metadata":{"id":"Neo5VCZVJoBl"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/mpmodelling_04.png'/>\n","</figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"xvXqKZZdZNjJ"},"source":["When your job is finished you will get an email with the corresponding fragments. You have to download them and place them inside the frags folder and make sure that their filename matches the one in XML RosettaScripts cell"]}]}