{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"WutUwTFo2e4j"},"source":["#Lab.11 / IBM3202 – Prediction of interactions from the coevolutionary analysis of sequence information"]},{"cell_type":"markdown","metadata":{"id":"FQMdCpigVK-X"},"source":["###Theoretical aspects\n","\n","As we have discussed, **natural protein sequences can vary highly** in terms of sequence identity and yet **retain their structure** and function. A striking example of this is the **AAA protein family of ATPases** (Fig. 1), present in all domains in life. All family members have a common AAA-cassette responsible for nucleotide binding and hydrolysis although **displaying sequence identity as low as 20%**. These results show that structure and function conservation can exist even in poorly conserved sequences. The same observations can be made for the analysis of **non-coding RNAs** that catalyze biochemical reactions and bind to specific ligands by **folding into complex three-dimensional structures**, such as in the case of ribozymes and riboswitches.\n"]},{"cell_type":"markdown","metadata":{"id":"mD5dj8UMSqEC"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/dca_01.png'/>\n","<figcaption>FIGURE 1. Architecture and sequence conservation of the AAA cassette, with its different domains shown as colored boxes, illustrating the low sequence identity of these proteins. Sequence patterns typical of the AAA family are listed in detail. In addition to the standard one-letter code for amino acids, the following symbols are used: <b>-</b>= D/E, <b>+</b>= R/K/H, <b>*</b>= charged , <b>O</b> = large and hydrophobic. Sequence colors denote conservation: red, green, black means almost absolutely conserved, well-conserved, and conserved, respectively. <br> Beyer A (1997) <i> Protein Sci 6(10), 2043-2058</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"P8b4ES6EHX9Z"},"source":["A **physical explanation** to this apparent paradox comes when considering that protein and non-coding RNA sequences are **one-dimensional bits of information** but these biomolecules themselves use all **four dimensions** – 3 spatial and 1 temporal – **to define their conformational space**. Yet, as we saw in our protein folding simulations, proteins do not explore the entirety of their possible conformational space (a.k.a. Levinthal’s paradox).\n","\n","For these biomolecules to efficiently and properly explore their conformational space towards their native structure, **specific (native) contacts between residues must be formed**. In physical terms, these may be hydrogen bonds, salt bridges, Van der Waals interactions, etc. Instead, we can simplify the definition of an interaction between residues by considering an **“interaction”** just as **two residues that have an effect on each other** and are therefore **constrained in the final structure**.\n","\n","Thus, the adequacy of protein and non-coding RNA sequences to fold into defined three-dimensional structures is not controlled by sequences themselves but controlled by their end product – the conformational space – and **how sequences evolve while keeping these stabilizing interactions between two or more residues in such sequence**."]},{"cell_type":"markdown","metadata":{"id":"xaWvTKKCDqfs"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/dca_02.gif'/>\n","<figcaption>Schematic representation of the residue pair contact information in three-dimensional structures that can be extracted from the coevolutionary analysis of protein and RNA sequences. The dotted boxes show how pairs of interacting residues are constrained to coevolve to maintain interactions that are crucial for the stability of the final structure.\n","Taken from GREMLIN, <i>gremlin.bakerlab.org/gremlin_faq.php</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"m6iYtcdxIU1l"},"source":["Then, a simple question arises: **can we infer these interactions from sequence information alone?** In the example above, we have a pair of interacting residues that occupy specific positions in a multiple sequence alignment and that can belong to a single or protein chains. As part of random drift, a deleterious substitution may have taken place, partially or completely **disrupting the interaction**, but this could be reversed by a **complementary mutation** that would **restore the interaction** through a different chemistry. This means that for a contact taking place in the native state of a protein, the **residues involved in such interaction can vary in identity** like any other, but they are **mutually restricted to keep identities compatible with such contact**. It ultimately means that the **interactions that are constrained in the final structure are encoded within sequence information as residues that coevolve**."]},{"cell_type":"markdown","metadata":{"id":"eiopP7EdVjqt"},"source":["##Experimental Overview"]},{"cell_type":"markdown","metadata":{"id":"xkpCp93eQsUS"},"source":["In this tutorial (which is largely based on the tutorial by Mehari B. Zerihun and deposited on [pyDCA's GitHub](https://github.com/KIT-MBS/pydca)) we will exemplify how we can **infer native contacts from the coevolutionary analysis of sequence information**. For a change, we will work on the coevolutionary analysis of **RNA**, instead of protein, sequences.\n","\n","We will particularly focus on comparing the experimental and predicted native interactions for the **adenosine deaminase (_add_) A-riboswitch**, one of the structurally simplest riboswitches and member of the Rfam family of **purine riboswitches**. Upon binding to adenine, the aptamer domain of _add_ A-riboswitch changes its folding pattern, which alters the conformation of its downstream expression platform and activates gene expression.\n","\n","For the coevolutionary analysis of RNA sequences we will employ the **Direct Coupling Analysis (DCA)** method. The DCA scores deliver quantitative information about the existence of physical contacts in a three-dimensional structure of a biomolecule, with top-scoring pairs accurately predicting native contacts that can be observed in experimentally solved protein and RNA structures."]},{"cell_type":"markdown","metadata":{"id":"buTRd47h4QfK"},"source":["#Part 0. Downloading and Installing the required software"]},{"cell_type":"markdown","metadata":{"id":"KSAoowL_4XIV"},"source":["Before we start, **remember to start the hosted runtime** in Google Colab.\n","\n","Then, we must install several pieces of software to perform this tutorial. Namely:\n","- **biopython** for manipulation of the PDB files\n","- **py3Dmol** for visualization of the protein structure.\n","- **infernal** (INFERence of RNA ALignment) for searching and aligning RNA sequences based on covariance models of sequence and RNA secondary structure consensus\n","- **pyDCA** for predicting native contacts through direct coupling analysis.\n","\n","**NOTE:** For the alignment of protein sequences, the recommended option would be **HMMER**, which enables searching and aligning protein sequences based on profile Hidden Markov Model (HMM) of position-specific conservation.\n","\n","**WARNING⚠️:** pyDCA requires more recent version of several python modules, and thus a **runtime restart** will be requested. Go ahead and accept it!\n"]},{"cell_type":"markdown","metadata":{"id":"7t7-UJnjOm8D"},"source":["1. We will first install infernal"]},{"cell_type":"code","metadata":{"id":"ba9bbMFk5ol3"},"source":["!apt-get install infernal"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["2. We need to downgrade python to be able to use pyDCA"],"metadata":{"id":"y_Dx8bu4Pcb_"}},{"cell_type":"code","source":["!apt-get install python3.7 python3-pip python3.7-distutils -y\n","!sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.7 1\n","!sudo update-alternatives --config python3"],"metadata":{"id":"tdg4H-7hPblQ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"x8zFzFh0OpPX"},"source":["3. Finally, we will install py3Dmol, pydca and biopython through `pip`.\n","  \n","  **DO NOT RESTART THE SESSION WHEN ASKED FOR IT, CLICK CANCEL**"]},{"cell_type":"code","metadata":{"id":"My3ObMLDY9G9"},"source":["!pip install biopython ipykernel pydca\n","!python3.10 -m pip install biopython py3Dmol"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"9GjXpNlbW-a3"},"source":["#Part I – Visualize the structure of _add_ A-riboswitch"]},{"cell_type":"markdown","metadata":{"id":"MoxY4Y_YXYFd"},"source":["We will first download and visualize the solved structure of _add_ A-riboswitch from _Vibrio fulnificus_ (PDB 1Y26) through biopython and py3Dmol.\n","\n","1. Download the structure of _add_ A-riboswitch as we have done in the past"]},{"cell_type":"code","metadata":{"id":"Xd7FJOtplOF0"},"source":["#Importing your PDB file using biopython\n","import os\n","from Bio.PDB import *\n","pdbid = ['1y26']\n","pdbl = PDBList()\n","for s in pdbid:\n","  pdbl.retrieve_pdb_file(s, pdir='.', file_format =\"pdb\", overwrite=True)\n","  os.rename(\"pdb\"+s+\".ent\", s+\".pdb\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gHQFin-HYxRt"},"source":["2. Visualize the structure using py3Dmol.\n","\n","**NOTE:** We are giving you the small exercise of indicating the chain ID for 1Y26 yourself."]},{"cell_type":"code","metadata":{"id":"-oGLjJmdYh1-"},"source":["import py3Dmol\n","#First we assign the py3Dmol.view as view\n","view=py3Dmol.view()\n","#The following lines are used to add the addModel class\n","#to read the PDB files\n","view.addModel(open('1y26.pdb', 'r').read(),'pdb')\n","#Here we set the background color as white\n","view.setBackgroundColor('white')\n","#Here we set the visualization style and color\n","view.setStyle({'chain':'X'},{'cartoon': {'colorscheme':'nucleic'}})\n","#You can activate the labels for each residue if you want\n","view.addResLabels({'resi':'13-83'},{'fontColor':'black','fontOpacity':1,'showBackground':'false'})\n","#Here we center the molecule for its visualization\n","view.zoomTo()\n","#And we finally visualize the structures using the command below\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Z65j3jScfS18"},"source":["**QUESTION❓:** Are all base-pairing interactions in this riboswitch of Watson-Crick type?"]},{"cell_type":"markdown","metadata":{"id":"ZJas1A0hffk-"},"source":["#Part II – Generate a multiple sequence alignment (MSA) of all known purine riboswitch family members using infernal"]},{"cell_type":"markdown","metadata":{"id":"p77vPJ1lgLEh"},"source":["As previously indicated, the _add_ A-riboswitch is a member of the family of **purine riboswitches** in the [**Rfam** database](http://rfam.xfam.org), a collection of RNA families. The benefit of Rfam is that it contains **seed alignments**, i.e. small MSA based on a subset of known sequences that is used to generate a **consensus covariance model** of sequence and structure conservation.\n","\n","These models are then used by Rfam to find **all possible RNA homologs for a given family**, and then all of these sequences for each RNA family are also **deposited in this database**.\n","\n","We will use these seed alignments and sequences of known family members as inputs for generating a covariance model-based MSA using infernal. The resulting MSA will be our input for our coevolutionary analysis, using the sequence of _add_ A-riboswitch as reference for the prediction of residue-pair contacts."]},{"cell_type":"markdown","metadata":{"id":"Rm4tR7SBueva"},"source":["1. We will first download the sequence of 1Y26 in FASTA format using biopython. A small change from previous scripts is that biopython downloads the sequence of interests in DNA alphabet. Thus, in order to obtain our sequence in RNA alphabet, we will create our own record with `SeqRecord` and use `transcribe()` to convert from DNA into RNA.\n","\n","**NOTE:** We are again giving you the small exercise of letting you set up the correct sequence ID for the RNA we want to download."]},{"cell_type":"code","metadata":{"id":"XXQ19ku5uipd"},"source":["import os\n","from pathlib import Path\n","from Bio import SeqIO, Entrez\n","from Bio.SeqRecord import SeqRecord\n","seqlist = ['1Y26_X'] # UPPERCASE LETTERS\n","for n in seqlist:\n","  #Setting up your email to be able to use Entrez\n","  Entrez.email = 'your.email@uc.cl'\n","  #Here, we set up a temporary handle with our downloaded sequence in fasta format\n","  temp = Entrez.efetch(db=\"sequences\",rettype=\"fasta\",id=n)\n","  #Creating a fasta file to write our downloaded sequence\n","  #Which is unfortunately downloaded as DNA\n","  seq_out = open(\"1Y26.fasta\",'w')\n","  #Reading the sequence information as a string in fasta format\n","  seq = SeqIO.read(temp, format=\"fasta\")\n","  print(\"Sequence is downloaded as DNA:\\n\" + seq.seq)\n","  #Transcribing the sequence, creating a new RNA record  and saving\n","  rnaseq = seq.seq.transcribe()\n","  print(\"We convert it back into RNA:\\n\"+rnaseq)\n","  rrec = SeqRecord(rnaseq, id=seq.id, description=\"reference\")\n","  SeqIO.write(rrec,seq_out,\"fasta\")\n","  #Closing both the temp handle and the FASTA file\n","  temp.close()\n","  seq_out.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"oO5pJY7kPfss"},"source":["2. Then, we will go to the [**Rfam** database](http://rfam.xfam.org) and search for the appropriate RNA family. You can easily do this by inputting _purine_ in the **_Jump To_** search box. This family is annotated as **Purine (RF00167)**.\n","\n","  Once here, we will obtain two different files:\n","  - Go to the **_Sequences_** menu to **download all sequences** in FASTA format, which we will use to generate a multiple sequence alignment (MSA)\n","  - Go to the **_Alignment_** menu and **download the seed alignment**, which is used for constructing a covariance model employed as profile for the MSA. You must download it in Stockholm format from the **_Formatting options_** menu.\n","\n","  Import both downloaded files by dragging them into the **Files** tab on the left sidebar of Google Colab."]},{"cell_type":"markdown","metadata":{"id":"n-DCxlNOSLYk"},"source":["3. We will now use the `cmbuild` module from infernal to generate our covariance model based on the Rfam seed alignment. This model will be stored as a `.cm` file"]},{"cell_type":"code","metadata":{"id":"DvkantIE5yrj"},"source":["!cmbuild RF00167.cm RF00167.stockholm.txt"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"lias6SkMSdSL"},"source":["4. Then, we will align all the downloaded sequences using our readily available covariance model as a MSA profile through the `cmalign` module of infernal. Please note that we are adding our reference sequence into the FASTA file to include it in our MSA. The resulting MSA will be stored as an **aligned FASTA** file named `RF00167_aligned.afa`.\n","\n","  Given that we are not interested in sequence deletions/insertions relative to the consensus model, we will include match columns in the output alignment with the `--matchonly` option."]},{"cell_type":"code","metadata":{"id":"n3tvFlL76GpT"},"source":["!gunzip RF00167.fa.gz\n","!cat RF00167.fa 1Y26.fasta > RF00167_full.fasta\n","!cmalign -o RF00167_aligned.afa --outformat AFA --matchonly RF00167.cm RF00167_full.fasta"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"rNMrLFU4U0Fr"},"source":["We have now obtained a MSA of all the known sequences of the purine riboswitch family (RF00167) of Rfam. As in our previous tutorials, you can visualize this alignment using [Alignment Viewer 2.0](https://fast.alignmentviewer.org/)"]},{"cell_type":"markdown","metadata":{"id":"pmx4F5DeU-fS"},"source":["#Part III – Coevolutionary analysis of purine riboswitches and accuracy of the prediction of residue pair contacts."]},{"cell_type":"markdown","metadata":{"id":"sdsh4ZSZVKVe"},"source":["Now, we will use pyDCA, a python implementation of DCA, to first predict residue pairs in contact in the three-dimensional structure of purine riboswitches based on sequence information alone, and then determine the accuracy of our prediction by establishing the number of true positives when compared to the contacts observed in the solved structure of _add_ A-riboswitch.\n","\n","1. First, we will trim our MSA file based on the lenght of the reference sequence, which in this case corresponds to the sequence of the aptamer of _add_ A-riboswitch that was solved by X-ray crystallography. The starting MSA is defined in the `rna_msa_file` path, whereas our sequence is defined in the `rna_refseq_file`. The trimmed MSA data to an output file `MSA_RF00167_Trimmed.fa` in FASTA format. The program installed is run as follows:\n","\n"," `pydca trim_by_refseq <biomolecule>  <alignment.fa>  <refseq_file.fa> --remove_all_gaps --verbose`\n","\n"," This will compare our target sequence to all the sequences in the alignment. Later, it will select the closest one and trim all the gap positions. The resulting msa will be generated in a folder named after the file with the \"Trimmed\" prefix.\n","\n"," **NOTE**: To trim protein sequences, we just need to change biomolecule='rna' to biomolecule='protein' and provide the respective MSA and reference sequence files."]},{"cell_type":"code","metadata":{"id":"RKZyDUE7vayJ"},"source":["!pydca trim_by_refseq rna RF00167_aligned.afa  1Y26.fasta --remove_all_gaps --verbose"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["2. We can now check that the top sequence is correctly trimmed"],"metadata":{"id":"pvgl_9WOTgTP"}},{"cell_type":"code","source":["!head Trimmed_RF00167_aligned/Trimmed_RF00167_aligned.fa"],"metadata":{"id":"M46OobtKTkpo"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"7leSC5tIXJ6Y"},"source":["3. Once done, we will compute our DCA scores using a **Pseudolikelihood maximization (plm)** algorithm. This  is different from the **mean-field (mf)** algorithm that we reviewed during the lectures. Analyzing this algorithm is outside the scope of this tutorial, it is elegantly described [in the BiorXiv of pyDCA](https://www.biorxiv.org/content/10.1101/805523v1).\n","\n","  For this analysis, we first create a plmDCA instance `plmdca_inst` to analyzed our trimmed RNA MSA. We also provide a series of optional parameters. The 2-body couplings $J_{ij}$ and single-site fields $h_i$ and $h_j$ for each position in the MSA are estimated by maximizing their pseudolikelihood from starting values through a gradient descent using an interative approach. The number of iterations is defined by `max_iterations`. Also, a maximum sequence identity percentage of 80% for all sequences in the MSA is established using the `seqid` parameter. Lastly, the number of threads is set to 2, the maximum for Google Colab.\n","\n","  After the fields and couplings are obtained, the DCA scores are computed from the Direct Information (DI):\n","<figure>\n","<center>\n","<img src='https://github.com/pb3lab/ibm3202/raw/master/images/dca_04.png'/>\n","</center>\n","</figure>\n","\n","  by calling the `compute_sorted_FN_APC()` method on `plmdca_inst`. This action returns the average product corrected (APC) DCA scores.\n","\n","  Here we will call the `plmdca` function with `compute_fn` function, which will calculate the Frobenius norm of the plmDCA, Average Product Corrected (APC) as we saw in the lectures.\n"]},{"cell_type":"code","metadata":{"id":"6oKZrPuyMUCA"},"source":["!plmdca compute_fn rna Trimmed_RF00167_aligned/Trimmed_RF00167_aligned.fa --apc --seqid 0.8 --num_threads 2 --max_iterations 500"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"mhWuPHSggJNY"},"source":["4. Let us print our top 10 site pairs and their corresponding DCA scores."]},{"cell_type":"code","metadata":{"id":"DCTZXKMr6BYf"},"source":["!head -n 22 PLMDCA_output_Trimmed_RF00167_aligned/PLMDCA_apc_fn_scores_Trimmed_RF00167_aligned.txt"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"LC9czoh-gPcL"},"source":["5. Once we have already obtained our DCA scores, we will use the DCA visualizer. This visualizer takes the type of biomolecule (`'rna'`), our reference sequence file (`refseq_file`) and our previously generated list of sorted DCA scores (`sorted_dca_scores`) as input to determine its residue pairs in contact based on sequence information alone.\n","\n","  When structural information is available, it also takes a PDB ID (`'1y26'`) and chain (`'x'`) as input parameters, as well as specified linear (i.e. sequence separation between residue pairs, `linear_dist`) and contact distances (`contact_dist`, in Å).This visualizer takes the following inputs:\n","\n","\n","\n","```\n","pydca plot_contact_map <biomolecule> <PDB_chain_name> <PDB_id/PDB_file.PDB> <refseq.fa> <DCA_file.txt> --verbose  \n","```\n","\n","**NOTE:** When we execute the above cell we see a  `You didn't supply RNA secondary structure file` warning. Since supplying an RNA secondary structure file is optional, we can ignore the warning."]},{"cell_type":"code","source":["!pydca plot_contact_map rna X 1Y26 1Y26.fasta\\\n"," PLMDCA_output_Trimmed_RF00167_aligned/PLMDCA_apc_fn_scores_Trimmed_RF00167_aligned.txt --verbose"],"metadata":{"id":"MwHYiEZrURrz"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ja5UOuOgiOe-"},"source":["6. We will now plot both contact maps, with the structure-based contacts on the upper left triangle, and the DCA-based contacts on the lower right triangle. Correctly predicted contacts are shown in green, whereas false positives are shown in red. Also note that the number of DCA contacts is equivalent to $L$, where $L$ is the lenght (i.e. number of columns) of the trimmed alignment."]},{"cell_type":"code","metadata":{"id":"U0NmcHC-MsTQ"},"source":["!awk 'NR>=23{print $5, $6, $1}' /content/contact_map_1Y26/contact_map1Y26.txt > contacts.txt\n","!echo X Y Z > /content/headers\n","!cat /content/headers /content/contacts.txt > /content/contacts2.txt\n","import numpy as np\n","import matplotlib.pyplot as plt\n","from matplotlib.colors import ListedColormap\n","\n","data = np.genfromtxt('/content/contacts2.txt', dtype=None, names=True, encoding=None)\n","x = data['X']\n","y = data['Y']\n","z = data['Z']\n","\n","# Define a dictionary to map categories to colors\n","category_colors = {\n","    'tp': 'green',\n","    'fp': 'red',\n","    'pdb': 'gray'\n","}\n","\n","# Create lists to store the inverted coordinates\n","inverted_x = []\n","inverted_y = []\n","colors = []\n","\n","plt.figure(figsize=(6, 6))\n","\n","for i in range(len(z)):\n","    category = z[i]\n","    if category in ('tp', 'fp'):\n","        inverted_x.append(y[i])  # Invert X and Y for 'tp' and 'fp' categories\n","        inverted_y.append(x[i])\n","        colors.append(category_colors[category])\n","    else:\n","        inverted_x.append(x[i])\n","        inverted_y.append(y[i])\n","        colors.append(category_colors[category])\n","\n","plt.scatter(inverted_x, inverted_y, c=colors, marker='o', s=5)  # You can change the 's' value as needed\n","for category, color in category_colors.items():\n","    plt.scatter([], [], c=color, label=category)\n","\n","plt.xlabel('Residue index')\n","plt.ylabel('Residue index')\n","plt.title('Contact map for the mfDCA prediction')\n","plt.legend(bbox_to_anchor=(1.05, 1.05))\n","plt.grid(True)\n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"qusYWvKOiulO"},"source":["7. Lastly, we will determine the accuracy of our coevolutionary analysis by looking at its true positive (TP) rate per rank. The TP rate per rank is the number of correctly predicted contacts per rank of the predicted pairs divided by all predictions at that rank"]},{"cell_type":"code","metadata":{"id":"w7FUXtQaJ3mw"},"source":["!pydca plot_tp_rate rna X 1Y26 1Y26.fasta\\\n"," PLMDCA_output_Trimmed_RF00167_aligned/PLMDCA_apc_fn_scores_Trimmed_RF00167_aligned.txt --verbose"],"execution_count":null,"outputs":[]},{"cell_type":"code","source":["import matplotlib.pyplot as plt\n","import pandas as pd\n","data = pd.read_csv('/content/TPR_1Y26/TPR_1Y26.txt', delimiter='\\t', skiprows=11, header=None, usecols=[0])\n","\n","data['Index'] = range(1, len(data) + 1)\n","data.set_index('Index', inplace=True)\n","plt.semilogx(data.index, data[0], linestyle='-')\n","plt.xlabel(\"Rank\")\n","plt.ylabel(\"TPR\")\n","plt.title(\"True positive rate for sorted APC\")\n","plt.show()"],"metadata":{"id":"ZrU4WWNyUk_n"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"vwt_fTB8jjuO"},"source":["In this resulting plot, the blue line corresponds to the TP rates for the predicted contacts, and the orange line corresponds to the theoretically maximum possible true positive rate for the contacts obtained from the PDB structure.\n","\n","**This is the end of the eleventh tutorial!** Good science!"]}]}